%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\documentclass[useAMS,usenatbib,referee]{biom}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newcommand{\bs}[1]{\boldsymbol{#1}}
\newcommand{\mb}[1]{\mathbf{#1}}
\newcommand*\red{\color{red}}

\usepackage[figuresright]{rotating}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{enumerate}
\usepackage{natbib}
\usepackage{url}
\usepackage{graphicx}
\usepackage{subfig}
\usepackage{booktabs}
%\usepackage{lipsum}
\usepackage{array}
\usepackage{float}
\usepackage{multirow}
\usepackage{appendix}
\usepackage{makecell}
\usepackage{color}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\setcounter{footnote}{2}

\title{Efficient Detection and Classification of Epigenomic Changes Under Multiple Conditions}

\author{Pedro L. Baldoni$^*$\email{baldoni@email.unc.edu},
Naim U. Rashid$^{**}$\email{naim@unc.edu}, and
Joseph G. Ibrahim$^{***}$\email{ibrahim@bios.unc.edu} \\
Department of Biostatistics, University of North Carolina at Chapel Hill, 
Chapel Hill, North Carolina, USA}

\begin{document}

\date{{\it Received May} 2020. {\it Revised May} 2020. {\it Accepted May} 2020.}

\pagerange{\pageref{firstpage}--\pageref{lastpage}}\pubyear{2020}

\volume{1}
\artmonth{May}
\doi{00.0000/0.000.0.0.0}

\label{firstpage}

\begin{abstract}
Epigenomics, the study of the human genome and its interactions with proteins and other cellular elements, has become of significant interest in recent years. Such interactions have been shown to regulate essential cellular functions and are associated with multiple complex diseases. Therefore, understanding how these interactions may change across conditions is central in biomedical research. Chromatin immunoprecipitation followed by massively-parallel sequencing (ChIP-seq) is one of several techniques to detect local changes in epigenomic activity (peaks). However, existing methods for differential peak calling are not optimized for the diversity in ChIP-seq signal profiles, are limited to the analysis of two conditions, or cannot classify specific patterns of differential change when multiple patterns exist. To address these limitations, we present a flexible and efficient method for the detection of differential epigenomic activity across multiple conditions. We utilize data from the ENCODE Consortium and show that the presented method, epigraHMM, exhibits superior performance to current tools and it is among the fastest algorithms available, while allowing the classification of combinatorial patterns of differential epigenomic activity and the characterization of chromatin regulatory states.
\end{abstract}

\begin{keywords}
%  Please place your key words in alphabetical order, separated
%  by semicolons, with the first letter of the first word capitalized,
%  and a period at the end of the list.
ChIP-seq; Differential Peak Call; Epigenomics; Hidden Markov Model; Mixture Model.
\end{keywords}

\maketitle

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}
\label{sec:introduction}

% Paragraph 1: Why does epigenetics matter? Nail down the problem 
Epigenomics, the study of the human genome and its interactions with proteins and other cellular elements, has become of significant interest in recent years. Such interactions have been shown to regulate essential cellular functions such as gene expression and DNA packaging \citep{kim2018polycomb}, resulting in downstream phenotypic impact. Therefore, the interrogation of how these interactions may change across conditions, such as cell types or treatments, is of marked interest in biomedical research. Several landmark articles have identified specific genomic regions of changing (differential) epigenomic activity between conditions as drivers of cell differentiation \citep{creyghton2010histone}, cancer progression \citep{varambally2002polycomb}, and a number of human diseases \citep{portela2010epigenetic}. Within differential regions, the delineation of specific patterns of change across conditions is also of interest, for example classifying the gain-of- or loss-of-activity in genomic loci due to treatment \citep{clouaire2014cfp1}. The identification of specific combinations of processes acting locally may also be informative, such as for segmenting the genome into regulatory states \citep{kundaje2015integrative}.

% Paragraph 3: Introduce ChIP-seq
To quantify local epigenomic activity on a genome-wide scale, a common high-throughput assay is chromatin immunoprecipitation followed by massively parallel sequencing (ChIP-seq). ChIP-seq experiments begin with cross-linking DNA and proteins within chromatin structures, which are then fragmented by sonication in a particular sample. DNA fragments bound to the protein of interest are isolated by chromatin immunoprecipitation, which are then sequenced via massively parallel high-throughput sequencing to generate short sequencing reads pertaining to the original fragments. Sequences are then mapped onto a reference genome through sequence alignment to determine their likely locations of origin. Genomic coordinates containing a high density of mapped reads, often referred to as enrichment regions (peaks), indicate likely locations of protein-DNA interaction sites, and all other regions are referred to as background regions. This local read density is often summarized by counting the number of reads mapped onto non-overlapping windows of fixed length tiling the genome (window read counts), forming the basis for downstream analyses. Across multiple conditions, regions exhibiting enrichment in at least one condition, but not across all conditions, indicate the presence of differential activity pertaining to the protein-DNA interaction of interest.

% Paragraph 4: Issues with the data/methods
To date, many differential peak callers (DPCs) have been proposed \citep{song2011identifying,stark2011diffbind,shen2013diffreps,chen2015novel,lun2015csaw,allhoff2016differential}. However, several challenges affect their ability to accurately detect regions of differential activity from the wide range of ChIP-seq experiments (Section \ref{sec:background}). First, differential regions may be both short or broad in length, causing difficulty for methods optimized for a particular type of signal profile \citep{stark2011diffbind,chen2015novel}. Second, methods that pool experimental replicates together \citep{song2011identifying} often exhibit more false positive calls compared to methods that jointly model replicates from each condition \citep{steinhauser2016comprehensive}. Third, the analysis of ChIP-seq data is often subject to complex biases that may vary across the genome, as differences in local read enrichment may depend on quantities such as the total read abundance in a given region or GC content \citep{teng2017accounting}. DPCs that disregard such complex effects or that solely rely on sample-specific global scaling factors or control subtraction methods \citep{stark2011diffbind,shen2013diffreps,chen2015novel,allhoff2016differential} may be prone to detecting spurious differences due to the lack of non-linear normalization methods \citep{lun2015csaw}. Reflecting these limitations, a recent comparison of DPCs demonstrated that current methods tend to detect either a large number of short peaks (low sensitivity) or exhibit a high number of false positive calls (low specificity) in ChIP-seq experiments with broad regions of enrichment \citep{steinhauser2016comprehensive}. Moreover, few methods are able to simultaneously test for differential activity across three or more conditions \citep{chen2015novel,lun2015csaw}, or can classify specific differential combinatorial patterns. Altogether, these limitations can impact the drawing of accurate insights from modern epigenomic studies.
 
% Paragraph 4: What we propose? Broad peaks, mixture model, and multiple conditions
Here, we propose an efficient and flexible statistical method to identify differential regions of enrichment from epigenomic experiments with diverse signal profiles and collected under common multi-replicate, multi-condition settings. Our method overcomes the limitations of current DPCs with three major features. First, it uses a hidden Markov model (HMM) to account for the diversity in differential enrichment profiles that may result from short and broad epigenomic ChIP-seq data sets. Second, it captures specific differential combinatorial patterns through a novel finite mixture model emission distribution within the HMM's differential state. Each mixture component pertains to a particular differential combinatorial pattern that is formed by the presence or absence of local enrichment across conditions, where a generalized linear model (GLM) is used to model the specific differential combinatorial pattern while accounting for sample- and window-specific normalization factors via offsets. Third, it enables the simultaneous detection and classification of epigenomic changes under three or more conditions, a novelty not yet available in any other DPC algorithm. The presented method offers additional benefits over current HMM-based DPC algorithms \citep{song2011identifying,allhoff2016differential} that include a GLM-based framework with an embedded mixture model, which allows the modeling of covariates of interest as well as the inclusion of model offsets for non-linear normalization, and a fast and accurate parameter estimation scheme via rejection-controlled EM algorithm (RCEM).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Data}
\label{sec:background}

% What are histones and why are they important
Histones are proteins that interact and condense  DNA in eukaryotic cells into structural units called nucleosomes. Multiple types of enzymatic modifications may be applied to histones, resulting in changes in local DNA packaging and chromatin accessibility mediated by nucleosomes \citep{bannister2011regulation}. In turn, cellular processes such as gene transcription, gene silencing, DNA repair, replication, and recombination are also affected. Proteins that interact with DNA and alter its functional properties are often referred to as epigenomic marks. For example, the trimethylation of histone H3 at lysines 36 and 27 (H3K36me3 and H3K27me3) are two types of histone modifications that tend to occur in genomic loci containing actively transcribed and repressed genes \citep{liu2016distinct}, respectively, and exhibit broad enrichment profiles. These marks have been investigated in cancer studies, where their absence is often observed in multiple cancer types \citep{wei2008loss}. As a result, H3K36me3 and H3K27me3 are considered to be key prognostic indicators in patients with breast, ovarian, and pancreatic cancer. The enhancer of zeste homolog 2 (EZH2), a major component of the polycomb complex PRC2 that catalyzes the methylation of H3K27me3 \citep{margueron2011polycomb}, is another example of a protein with experimental signal characterized by broad enrichment domains and co-occurs with the activity of H3K27me3.

Using ChIP-seq data pertaining to histone modifications H3K27me3 and H3K36me3, and to EZH2 from the ENCODE Consortium, we find that current DPCs have difficulty in accurately detecting broad regions of differential enrichment between several common cell lines (Figure~\ref{fig:fig_motivation}). In line with previous findings \citep{steinhauser2016comprehensive}, we observe that even current DPCs designed for broad data \citep{song2011identifying,allhoff2016differential} tend to detect either overly fragmented differential peaks or call regions exhibiting no difference in experimental signal between conditions as differential (Figure~\ref{fig:fig_motivation}A). The low specificity and sensitivity of such methods may impair the biological interpretation of the resulting peak calls in downstream analyses. Methods that rely on candidate peaks may also exhibit a compromised performance due to the limitations of single-sample peak callers in broad data \citep{stark2011diffbind,chen2015novel}. In addition, most current DPCs restrict their application to the analysis of two experimental conditions. For methods that are tailored for the analysis of three or more conditions, the classification of specific differential combinatorial patterns across conditions (or across various epigenomic processes) is still an open problem. The classification of such patterns would allow researchers to, for example, quantify treatment responses on the epigenomic level \citep{clouaire2014cfp1}, or identify sets of processes working together to regulate local chromatin state. We find that the performance of such methods exhibit low sensitivity and specificity in calling differential regions in broad marks (Figure~\ref{fig:fig_motivation}B).
\begin{figure}
\begin{center}
\includegraphics[scale = 0.75]{../../Main/Figure_epigraHMM/Figure_epigraHMM.pdf}
\end{center}
\caption{At the top, differential peak calls from current methods (under false discovery rate control of 0.05) for the histone modification H3K27me3 among cell lines Helas3, Hepg2, and Huvec. Shaded regions indicate  observed differential enrichment, and each vertical line type bordering each region represents a different combinatorial pattern of enrichment across cell lines. Optimal methods would call broad peaks inside shaded regions and no peaks outside them. At the bottom, general overview of the presented method for simultaneous detection and classification of broad and short differential regions of enrichment. This figure appears in color in the electronic version of this article.
\label{fig:fig_motivation}}
\end{figure}

We assessed the performance of our model on ChIP-seq experiments characterized by broad peaks (H3K36me3, H3K27me3, and EZH2) and short peaks (H3K27ac, H3K4me3, and the transcription factor CTCF). In simulations (Section \ref{sec:simulation}) and in data sets from the ENCODE Consortium \citep{landt2012chip}, we show that our model addresses the issues of the current peak callers in broad data (Section \ref{sec:broad}), while being flexible for short peaks (Section \ref{sec:sharp}) and comparable to the fastest DPCs regarding the computation time. We show that our method can also be utilized for genomic regulatory state segmentation when studying multiple types of epigenomic processes from a single condition or cell line (Section \ref{sec:predpostprob}). The Web Appendices A1 and A2 present the data accession codes and the data pre-processing steps, respectively. Code implementing the method and to replicate the presented results are available in the Web Appendices A3 and A4, respectively.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Methods}
\label{sec:methods}

\subsection{Statistical Model}
\label{subsec:hmmmixt}
Let $Y_{hij}$ denote the random variable pertaining to the ChIP read count for genomic window $j$ from sample $i$ of condition $h$, where $j=1,\ldots,M$, $i=1,\ldots,n_{h}$, $h=1,\ldots,G$, and let $y_{hij}$ be the observed count. Here, $n_h$ is the number of samples in condition $h$ and $N = \sum_{h = 1}^G n_h$ is the total number of samples across the $G$ conditions.  At the $j^{th}$ window, let $\mb{y}_{..j}=(y_{11j},\ldots,y_{Gn_{G}j})^\prime$ denote the $N \times 1$ vector of ChIP window read counts across all samples and conditions, and let $\mb{y}=(\mb{y}_{..1},\ldots,\mb{y}_{..M})$ denote the corresponding $N \times M$ matrix of window read counts spanning all windows, samples, and conditions. We assume that each window belongs to one of three possible hidden states: consensus background (state 1), differential (state 2), and consensus enrichment (state 3). Windows exhibiting low (high) enrichment across all conditions will be modeled by an emission distribution pertaining to the consensus background (enrichment) state. Windows exhibiting enrichment under at least one condition, but not all conditions, will be modeled by an emission distribution pertaining to the differential state. If $G$ conditions are of interest, there are $L = 2^G-2$ possible differential combinatorial patterns of enrichment and background across conditions at a given window. The emission distribution pertaining to the differential state models all $L$ possible differential combinatorial patterns via a mixture model with mixture proportions $\bs{\delta} = (\delta_{1},\ldots,\delta_{L})^\prime$, such that $\sum_{l=1}^{L}\delta_{l}=1$ (Web Figure 1).

To model transitions between states, we assume a single latent discrete time stationary Markov chain $\mb{Z}=\{Z_{j}\}_{j=1}^{M}$, $Z_{j}\in\{1,2,3\}$, with state-to-state transition probabilities $\bs{\gamma}=(\gamma_{11},\gamma_{12},\ldots,\gamma_{33})^\prime$ and initial probabilities $\bs{\pi}=(\pi_{1},\pi_{2},\pi_{3})^\prime$, such that $\sum_{s=1}^{3}\gamma_{rs}=1$ and $\sum_{s=1}^{3}\pi_{s}=1$ for $r\in\{1,2,3\}$.  To facilitate the notation, let $f_{r}(\mb{y}_{..j}| \bs{\psi}_{r})$ denote the emission distribution corresponding  to the $r^{th}$ hidden state, where $\bs{\Psi} = (\bs{\pi}^\prime,\bs{\gamma}^\prime,\bs{\delta}^\prime,\bs{\psi}^\prime)^\prime$ denotes the vector of all model parameters, $\bs{\psi}=(\bs{\psi}_{1}^\prime,\bs{\psi}_{2}^\prime,\bs{\psi}_{3}^\prime)^\prime$ denotes each state's set of emission distribution-specific parameters, and $\mathcal{Z}$ denotes the set of $3^M$ possible state paths of length $M$. Then, the likelihood function pertaining to the proposed HMM may be written as
\begin{align}\label{eq:difflik}
f(\mathbf{y}|\mathbf{x};\bs{\Psi}) &= \sum_{\mb{Z}\in\mathcal{Z}}\left\{\prod_{r=1}^{3}\pi_{r}^{I(Z_{1}=r)}\times\left(\prod_{j=2}^{M}\prod_{r=1}^{3}\prod_{s=1}^{3}\gamma_{rs}^{I(Z_{j-1}=r,Z_{j}=s)}\right)\right.\times\\
&\times\left.\left(\prod_{j=1}^{M}f_{1}\left(\mb{y}_{..j}|\bs{\psi}_{1}\right)^{I(Z_{j}=1)}f_{2}\left(\mb{y}_{..j}|\mb{x};\bs{\delta},\bs{\psi}_{2}\right)^{I(Z_{j}=2)}f_{3}\left(\mb{y}_{..j}|\bs{\psi}_{3}\right)^{I(Z_{j}=3)}\right)\right\}.\nonumber
\end{align}
Here, $\mb{x}$ is a fixed $G \times L$ design matrix enumerating each of the $L$ possible differential combinatorial patterns in terms of the presence or absence of enrichment across each of the $G$ conditions, only in the emission distribution of the differential state.

We assume that read counts pertaining to genomic windows from the consensus background ($r=1$) and consensus enrichment  ($r=3$) states follow a Negative Binomial (NB) distribution with state-specific parameters $\bs{\psi}_{r}=(\mu_{(r,hij)},\phi_{r})^\prime$, with mean $\mu_{(r,hij)}$ and variance $\mu_{(r,hij)}(1+\mu_{(r,hij)}/\phi_{r})$. Assuming independence of read counts across experiments and samples, conditional upon the HMM state, the emission distribution of the consensus background and consensus enrichment states, respectively, can be written as
\begin{align}\label{eq:nondiffdensity}
f_{r}(\mb{y}_{..j}|\bs{\psi}_{r}) &= \prod^{G}_{h=1}\prod^{n_{h}}_{i=1}\frac{\Gamma(y_{hij}+\phi_{r})}{y_{hij}!\Gamma(\phi_{r})}\left(\frac{\phi_{r}}{\mu_{(r,hij)}+\phi_{r}}\right)^{\phi_{r}}\left(\frac{\mu_{(r,hij)}}{\mu_{(r,hij)}+\phi_{r}}\right)^{y_{hij}},\quad r\in\{1,3\},
\end{align}
with $y_{hij}\in\{0,1,2,\ldots\}$, such that $\log(\mu_{(1,hij)})=\beta_{1} + u_{hij}$, $\log(\phi_{1})=\lambda_{1}$, $\log(\mu_{(3,hij)})=\beta_{1} + \beta_{3} + u_{hij}$, and $\log(\phi_{3})=\lambda_{1} + \lambda_{3}$. The offset $u_{hij}$ adjusts for technical artifacts and allows the non-linear normalization of the signal profile across genomic windows, conditions, and samples (Web Appendix B2). When $u_{hij}=0$, $\beta_{1}$ and $\lambda_{1}$ represent the log-mean and log-dispersion, respectively, of read counts pertaining to consensus background state windows, whereas $\beta_{3}$ and $\lambda_{3}$ represent the difference in log-mean and log-dispersion of read counts from consensus enrichment state windows relative to consensus background state windows.

For windows belonging to the differential state ($r = 2$), we assume that the corresponding read counts are modeled by a $L$-component finite mixture model with mixture components that follow a Negative Binomial distribution, where each component corresponds to a particular differential combinatorial pattern. To define these patterns, let us consider the sets $S_{1},\ldots,S_{L}$ that delineate the subset of the $G$ conditions that are enriched in each of the $L$ differential combinatorial patterns. For instance, if $G=3$, the sets $S_{1} =\{1\}$, $S_{2} =\{2\}$, $S_{3} =\{3\}$, $S_{4} =\{1,2\}$, $S_{5} =\{1,3\}$, and $S_{6} =\{2,3\}$ define the six possible differential combinatorial patterns of enrichment and background across three conditions. That is, the set  $S_{1}$ denotes enrichment in only the first condition and background in all others, whereas the set $S_{6}$ denotes enrichment in conditions 2 and 3 and background in condition 1. The presence or absence of enrichment in each of the $L$ sets is encoded into each column of $\mb{x}=(\mb{x}_{1},\ldots,\mb{x}_{L})$, such that $\mb{x}_{l} = (x_{1l},\ldots,x_{Gl})'$, and $x_{hl} = I(h\in S_{l})$ for $l=1,\ldots,L$ and $h = 1,\ldots,G$. That is, $\mb{x}_{l}$ is the $G \times 1$ vector of binary indicator variables denoting which subset of conditions are enriched in pattern (mixture component) $l$. A graphical illustration of our proposed model is provided in Web Figure 1.

Let $\bs{\psi}_{2}$ denote the state-specific parameter vector pertaining to the differential state and let $\bs{\psi}_{(2,l)}$ denote the set of parameters pertaining to the $l^{th}$ mixture component.  Assuming independence of read counts across conditions and samples, conditional upon the differential HMM state, the finite mixture model emission distribution can be written as
\begin{align}\label{eq:diffdensity}
f_{2}(\mb{y}_{..j}|\mb{x};\bs{\delta},\bs{\psi}_{2}) &= \sum_{l=1}^{L}\delta_{l}\left[\prod_{h=1}^{G}\prod_{i=1}^{n_{h}}\frac{\Gamma\left(y_{hij}+\phi_{(2,l,h)}\right)}{y_{hij}!\Gamma\left(\phi_{(2,l,h)}\right)}\left(\frac{\phi_{(2,l,h)}}{\mu_{(2,l,hij)}+\phi_{(2,l,h)}}\right)^{\phi_{(2,l,h)}}\times\right.\nonumber\\
&\left.\times\left(\frac{\mu_{(2,l,hij)}}{\mu_{(2,l,hij)}+\phi_{(2,l,h)}}\right)^{y_{hij}}\right],\quad y_{hij}\in\{0,1,2,\ldots\}, % &= \sum_{l=1}^{L}\delta_{l}f_{(2,l)}\left(\mb{y}_{..j}|\mb{x}_l;\bs{\psi}_{(2,l)}\right),\nonumber\\
\end{align}
where $\mu_{(2,l,hij)}$ and $\phi_{(2,l,h)}$ are the mean and dispersion, respectively, pertaining to read counts originating from window $j$ and sample $i$ in condition $h$ from the mixture component $l$. We assume that $\log(\mu_{(2,l,hij)}) = \beta_{1} + \beta_{3}x_{hl} + u_{hij}$ and $\log(\phi_{(2,l,h)}) = \lambda_{1} + \lambda_{3}x_{hl}$. That is, in the mixture component $l$, we utilize the same consensus background (consensus enriched) log-mean and log-dispersion from (\ref{eq:nondiffdensity}) in all conditions that are specified by $\mb{x}_{l}$ to be background (enriched) in the $l^{th}$ differential combinatorial pattern. There are several advantages to such a parametrization for the differential emission distribution. For example, it ensures that windows exhibiting differential enrichment across conditions share means and dispersions that are common between the consensus background and consensus enrichment states, a reasonable assumption that significantly increases computational efficiency.  Utilizing a mixture model as the differential state emission distribution avoids the computational burden that would come from assuming separate hidden states for each of the $L$ differential combinatorial patterns, particularly as $G$ increases.  We evaluate the strength of these assumptions through multiple simulations and a real data benchmarking analysis in Sections \ref{sec:simulation} and \ref{sec:realdata}.

Two novel features result from our proposed approach that are relevant to the context of differential enrichment detection from ChIP-seq experiments. By using a modified version of the Expectation-Maximization (EM) algorithm to estimate the model parameters, we are able not only to detect differential enrichment regions across multiple conditions,  but we can also classify various differential combinatorial patterns of enrichment within broad and short differential enrichment domains. With state-specific parameters, the current implementation of the method allows the direct modeling of continuous covariates (e.g. input controls; Web Appendix A3), for which a state-level testing of their effects on the read count distribution could be performed. In a simulation study and in real data analyses, however, we did not observe a significant improvement in performance in differential peak detection after accounting for the effect of input controls (Web Appendix B2), a fact that has also been observed by others \citep{lun2015csaw}.

\subsection{Estimation}
\label{subsec:hmmest}
To simplify the parameter estimation in (\ref{eq:obslik}), we introduce another set of latent variables $\mb{W}=(\mb{W}_{1}^\prime,\ldots,\mb{W}_{M}^\prime)^\prime$, such that $\mb{W}_{j}=(W_{j1},\ldots,W_{jL})^\prime$ for $j=1,\ldots,M$. We assume that $\mb{W}$ is a sequence of independent random vectors such that $\mb{W}_{j}|(Z_{j}=2)\sim\text{Multinomial}(1,\bs{\delta})$ and $\mb{W}_{j}|(Z_{j}=r)=0$ with probability 1 if $r=\{1,3\}$. Under this setup, one may define the data generating mechanism when $Z_{j}=2$ (differential state) and $W_{jl}=1$ ($l^{th}$ differential combinatorial pattern) such that read counts pertaining to genomic window $j$ are sampled from $f_{(2,l)}$ given $\bs{\psi}_{(2,l)}$ and $\mb{x}_{l}$. Let $\mathcal{W}$ denote the set of $L^M$ possible combinations of latent vectors $\mb{W}$. Hence, the likelihood function of the observed data (\ref{eq:difflik}) can be rewritten as
\begin{align}\label{eq:obslik}
f(\mathbf{y}|\mathbf{x};\bs{\Psi}) &= \sum_{\mb{Z}\in\mathcal{Z}}\sum_{\mb{W}\in\mathcal{W}}\left\{\left[\prod_{r=1}^{3}\pi_{r}^{I(Z_{1}=r)}\prod_{j=2}^{M}\prod_{r=1}^{3}\prod_{s=1}^{3}\gamma_{rs}^{I(Z_{j-1}=r,Z_{j}=s)}\right.\right]\times\left[\prod_{j=1}^{M}\left(\prod_{l=1}^{L}\delta_{l}^{W_{jl}}\right)^{I(Z_{j}=2)}\right]\times\nonumber\\
&\times \left.\left[\prod_{j=1}^{M}f_{1}\left(\mb{y}_{..j}|\bs{\psi}_{1}\right)^{I(Z_{j}=1)}\left(\prod_{l=1}^{L}f_{(2,l)}(\mb{y}_{..j}|\mb{x}_l;\bs{\psi}_{(2,l)})^{W_{jl}}\right)^{I(Z_{j}=2)}f_{3}\left(\mb{y}_{..j}|\bs{\psi}_{3}\right)^{I(Z_{j}=3)}\right]\right\}, %&= \sum_{\mb{Z}\in\mathcal{Z}}\left\{\left[\sum_{\mb{W}\in\mathcal{W}}f(\mb{y}|\mb{w},\mb{z},\mb{x};\bs{\psi})f(\mb{w}|\mb{z},\mb{x};\bs{\delta})\right]f(\mb{z}|\mb{x};\bs{\pi},\bs{\gamma})\right\},\\
\end{align}
where $f_{(2,l)}(\mb{y}_{..j}|\mb{x}_l;\bs{\psi}_{(2,l)})$ is defined as in (\ref{eq:diffdensity}). In the $t^{th}$ iteration of the EM algorithm, the $Q$ function of the complete data log-likelihood can be written as
\begin{align}\label{eq:diffQfunc}
Q\left(\bs{\Psi}|\bs{\Psi}^{(t)}\right) &= E_{\mb{Z}}\left(E_{\mb{W}|\mb{Z}}\left(\log\left(f(\mb{y},\mb{W},\mb{Z}|\mb{x};\bs{\Psi})\right)|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right)|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right),\nonumber\\
&= Q_{0}(\bs{\pi},\bs{\gamma}|\bs{\Psi}^{(t)})+Q_{1}(\bs{\psi}_{1}|\bs{\Psi}^{(t)})+Q_{2}(\bs{\delta},\bs{\psi}_{2}|\bs{\Psi}^{(t)})+Q_{3}(\bs{\psi}_{3}|\bs{\Psi}^{(t)}),
\end{align}
where $Q_{0}(\bs{\pi},\bs{\gamma}|\bs{\Psi}^{(t)})$, $Q_{1}(\bs{\psi}_{1}|\bs{\Psi}^{(t)})$, $Q_{2}(\bs{\delta},\bs{\psi}_{2}|\bs{\Psi}^{(t)})$, and $Q_{3}(\bs{\psi}_{3}|\bs{\Psi}^{(t)})$ are defined in Web Appendix B5. In the E-step of the EM algorithm, we compute the posterior probabilities from (\ref{eq:diffQfunc}). The quantities $Pr\left(Z_{j}=r|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right)$ and $Pr\left(Z_{j-1}=r,Z_{j}=s|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right)$, defined in Web Appendix B5, can be calculated through the Forward-Backward algorithm and $Pr(W_{jl}=1|Z_{j}=2,\mb{y}_{..j},\mb{x};\bs{\Psi}^{(t)})=f_{(2,l)}(\mb{y}_{..j}|\mb{x}_l;\bs{\psi}_{(2,l)}^{(t)})\delta_{l}^{(t)}/\sum_{k=1}^{L}f_{(2,k)}(\mb{y}_{..j}|\mb{x}_k;\bs{\psi}_{(2,k)}^{(t)})\delta_{k}^{(t)}$ for $l=1,\ldots,L$.

The $Q$ function is maximized with respect to the parameters $\bs{\Psi}=(\bs{\pi}^\prime,\bs{\gamma}^\prime,\bs{\delta}^\prime,\beta_{1},\beta_{3},\lambda_{1},\lambda_{3})^\prime$ during the M-step of the algorithm. Estimates for the initial and transition probabilities can be directly calculated as $\hat{\pi}^{(t+1)}_{r}=Pr\left(Z_{1}=r|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right)$ and $\hat{\gamma}^{(t+1)}_{rs}=\sum_{j=2}^{M}Pr(Z_{j-1}=r,Z_{j}=s|\mb{y},\mb{x};\bs{\Psi}^{(t)})/\sum_{j=2}^{M}Pr(Z_{j-1}=r|\mb{y},\mb{x};\bs{\Psi}^{(t)})$, respectively, restricted to $\sum_{r=1}^{3}\hat{\pi}^{(t+1)}_{r}=1$ and $\sum_{s=1}^{3}\hat{\gamma}^{(t+1)}_{rs} = 1$, for $r\in\{1,2,3\}$. We perform conditional maximizations to compute estimates of the remaining model parameters $(\bs{\delta}^\prime,\beta_{1},\beta_{3},\lambda_{1},\lambda_{3})^\prime$. First, mixture proportions can be estimated as $\hat{\delta}_{l}^{(t+1)}=\sum_{j=1}^{M}Pr(Z_{j}=2|\mb{y},\mb{x};\bs{\Psi}^{(t)})Pr(W_{jl}=1|Z_{j}=2,\mb{y}_{..j},\mb{x};\bs{\Psi}^{(t)})/\sum_{j=1}^{M}Pr(Z_{j}=2|\mb{y},\mb{x};\bs{\Psi}^{(t)})$. Estimating $(\beta_{1},\beta_{3},\lambda_{1},\lambda_{3})^\prime$ from (\ref{eq:diffQfunc}) can be seen as obtaining parameter estimates from a series of weighted NB regression models with shared mean and dispersion parameters (Web Appendix B5). We jointly estimate these quantities via the algorithm BFGS \citep{fletcher2013practical}.

The estimation scheme is robust to situations where certain differential combinatorial patterns of enrichment are rare (Figure \ref{fig:fig_multimarks}). This unique characteristic results from the fact that ChIP-seq experiments often provide enough data (usually $M>10^7$ non-overlapping windows of 250 bp fixed size for the human reference genome) to estimate the parameters $(\beta_{1},\beta_{3},\lambda_{1},\lambda_{3})^\prime$, which are shared across all $L$ mixture components and HMM states. If pruning differential combinatorial patterns of the differential mixture component is of interest, the optimal number of mixture components $L^{*}$, $L^{*}<L$, can be selected via the Bayesian Information Criterion (BIC) for HMMs. We observed that selecting the optimal number of mixture components based on BIC agrees with the pruning of rare differential combinatorial patterns that we would not biologically expect to observe in real data (Web Appendix B3).

To obtain the parameter estimates $\hat{\bs{\Psi}}$, the EM algorithm iterates until the maximum absolute relative change in the parameter estimates three iterations apart is less than $10^{-3}$ for three consecutive iterations. To reduce the computation time, we make use of a RCEM algorithm with threshold 0.05. Briefly, the RCEM algorithm substantially reduces the dimensionality of the data during the M-step by randomly assigning a zero posterior probability to genomic windows unlikely to belong to each of the HMM states. The current estimation set up allows genomic windows exhibiting equal distribution of read counts to have their posterior probability aggregated during the M-step of the algorithm. Often, the distribution of read counts along the genome is highly concentrated on a particular set of values, such as 0, 1, and 2 for instance. Genomic windows exhibiting a particular pattern of counts across samples and conditions can have their posterior probability aggregated during the M-step, which further reduces the dimensionality of the objective function during the numerical optimization and leads to a fast gradient-based optimization. %Genomic windows with assigned zero posterior probability due to the RCEM approach do not contribute to the final set of aggregated posterior probabilities.

Once the algorithm reaches convergence, the final set of HMM posterior probabilities can be used to segment the genome into consensus background, differential, or consensus enrichment windows. Approaches that control the total false discovery rate (FDR) via posterior probabilities \citep{efron2001empirical} or that estimate the most likely sequence of hidden states \citep{viterbi1967error} can be used for such purposes. Let $\hat{\rho}_{j2} = Pr\left(Z_{j}=2|\mb{y},\mb{x};\hat{\bs{\Psi}}\right)$ denote the estimated posterior probability that the $j^{th}$ genomic window belongs to the differential HMM state, $j=1,\ldots,M$. For a cutoff of posterior probability $\alpha$, the nominal FDR is $\sum_{j=1}^{M}(1-\hat{\rho}_{j2})I(\hat{\rho}_{j2}\geq 1-\alpha)/\sum_{j=1}^{M}I(\hat{\rho}_{j2}\geq 1-\alpha)$, where $I(\cdot)$ is an indicator function. The posterior probability cutoff is then chosen by controlling the total FDR. Differential regions of enrichment are formed by merging adjacent windows that either meet a nominal FDR threshold level for the differential HMM state or belong to the same Viterbi's predicted state. A discussion about these two approaches under the proposed model is presented in the Web Appendix B4. Additional details of proposed EM algorithm and the implemented code are available in the Web Appendices B5 and A3, respectively.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Simulation Studies}
\label{sec:simulation}
We evaluate the presented model in two independent simulation studies of broad epigenomic marks. In the first study (Section~\ref{sec:sim1}), we simulated read count-based data to assess the precision of the parameter estimation scheme, the performance of differential peak detection, and the accuracy of the classification of specific differential combinatorial patterns of enrichment within differential peaks. In the second simulation study (Section~\ref{sec:sim2}), we utilize the simulation pipeline presented in \cite{lun2015csaw} to generate synthetic ChIP-seq reads from \textit{in silico} experiments with broad differential peaks. The aim of the second simulation study was to compare our model with other DPCs in a more realistic scenario with broad peaks, while also avoiding the assumption of a parametric model for the data.

\subsection{Read Count Simulation}
\label{sec:sim1}
Read counts were simulated under different scenarios that varied regarding the type of histone modification mark (H3K36me3 and H3K27me3), number of genomic windows ($M$, $10^5$, $5\times 10^5$, and $10^6$ windows), number of conditions ($G$, 2, 3, and 4), and number of replicates per condition ($n$, 1, 2, and 4). We further assessed our model under different Signal-to-Noise Ratio (SNR) levels. We define the SNR as the ratio between the means of consensus enrichment and consensus background emission distributions. Mean and dispersion parameters used in this simulation study were estimated from ENCODE data and are presented in the Web Appendix C1 for all the scenarios. Different SNR levels were defined by decreasing the ratio of the means in decrements of 10\% while maintaining the mean-variance relationship. Read counts were assumed to follow a NB distribution and were simulated using a first-order Markov chain with $2^G$ states, representing every combination of background and enrichment across $G$ conditions. We aimed to assess whether our model was able to assign all $2^G-2$ simulated differential states to the differential HMM state, while maintaining a precise parameter estimation scheme and accurate classification of differential combinatorial patterns.

\subsubsection{Simulation Results}
Table~\ref{Table:table_simulation} shows the true values and the average relative bias of parameter estimates (and the $2.5^{th}$, $97.5^{th}$ percentiles) from a hundred simulated data sets relative to the scenario of H3K27me3 with $10^5$ genomic windows (see Web Appendix C1 for additional details). Results are shown for different levels of SNR, number of conditions, and number of replicates per condition. Overall, no significant differences regarding the relative bias of parameter estimates were observed across simulations under different number of genomic windows. Depending on the number of conditions, the observed relative bias and the range of the reported percentiles tended to decrease as more replicates were included in the analyses. This effect was particularly significant in scenarios with four conditions with respect to parameters $\beta_{3}$ and $\lambda_{3}$. In general, scenarios with higher SNR showed lower relative bias and variability of the parameter estimates in comparison to scenarios with lower SNR, regardless of the number of conditions or replicates per condition. In scenarios with lower SNR levels or higher number of conditions, these results also highlight the importance of experimental replicates to achieve precise parameter estimates. The proposed estimation approach via EM algorithm led to precise parameter estimates and was robust to a data generating mechanism that was different than the one assumed by the proposed model.
\begin{center}
\input{../../Main/Table_Simulation_ReadCounts/Table_Simulation_ReadCounts.tex}
\end{center}

\subsection{Sequencing Read Simulation}
\label{sec:sim2}
We performed a second simulation study aiming to compare the proposed model with the current DPCs ChIPComp, csaw, DiffBind, diffReps, RSEG, and THOR. We used the simulation pipeline presented by \cite{lun2015csaw} where data were generated in a more general scheme without a particular read count model assumption. Here, sequencing reads from broad ChIP-seq experiments were generated for two conditions and two replicates per condition. For the differential peaks callers ChIPComp and DiffBind that require sets of candidate regions, we followed the analyses presented by \cite{lun2015csaw} and called peaks in advance using HOMER. Peaks were then used as input in the respective software for differential call. A hundred simulated data sets were generated and peaks were called by all the methods under multiple nominal FDR thresholds. For our method and RSEG, window-based posterior probabilities were used to control the total FDR as described in Section \ref{subsec:hmmest}.% We refer to the method presented in this article as epigraHMM. We attempted to utilize MACS in this simulation study but the peak caller failed to output peaks for the simulated data.

\subsubsection{Simulation Results}
Figure~\ref{fig:fig_simulation_csaw} shows the main results of our second simulation study. Out of 100 simulated data sets, RSEG either failed to analyze the data due to internal errors or called the entire genome as differential in 26 and in 3 instances, respectively. Problematic cases from RSEG were not considered in this simulation study. Similar issues have been previously reported in other studies \citep{starmer2016detecting}. We observed that our method showed the highest observed sensitivity among all DPCs, regardless of the nominal FDR thresholding level, while maintaining a moderate observed FDR (Figure~\ref{fig:fig_simulation_csaw}A). Here, the observed FDR is the proportion of genomic windows incorrectly called as differential out of the total number of windows called as differential. Methods such as diffReps, RSEG, and THOR showed higher observed FDR levels than the nominal threshold due to the excessive number of differential peaks called outside true differential regions (shaded area in Figure~\ref{fig:fig_simulation_csaw}D). While diffReps and THOR called an excessive number of short and discontiguous peaks, RSEG called regions that were usually wider than the observed differential enrichment regions. These results are further illustrated in Figure~\ref{fig:fig_simulation_csaw}B, where we present the average ratio of the number of called and simulated peaks (y-axis) and the average number of called peaks intersecting true differential regions (x-axis). Regarding the computation time, the HMM-based algorithms RSEG and THOR appeared to be the most computationally intensive and required longer amounts of time to analyze the data. In Figure~\ref{fig:fig_simulation_csaw}C, we present the box plots of computing time (in minutes) across a hundred simulated data sets for all benchmarked methods. While still being an HMM-based algorithm, our method was among the fastest tools for differential peak detection due to the implemented strategies to improve the computation time of the EM algorithm (Section \ref{subsec:hmmest}). Figure~\ref{fig:fig_simulation_csaw}D shows an example of a genomic region with simulated data and called peaks from various methods using nominal FDR threshold 0.05. As shown, our method was able to consistently cover most of true differential regions with broad peaks while exhibiting a limited number of false discoveries (see Web Appendix C2 for additional results).
\begin{figure}
\begin{center}
\includegraphics[scale = 0.65]{../../Main/Figure_Simulation_SequencingReads/Figure_Simulation_SequencingReads.pdf}
\end{center}
\caption{Sequencing read-based simulation from the csaw pipeline. (A): average observed sensitivity and FDR for various methods. (B): scatter plot of average ratio of called and simulated peaks (y-axis) and number of called peaks intersecting true differential regions (x-axis). (C): box plot of computing time (in minutes) for various algorithms. (D): an example of differential peak calls under a nominal FDR control of 0.05. Shaded areas indicate true differential peaks. This figure appears in color in the electronic version of this article.
\label{fig:fig_simulation_csaw}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Application to ENCODE Data}
\label{sec:realdata}
We applied our method to ChIP-seq data from the ENCODE Consortium (Section \ref{sec:background}) to detect differential regions of enrichment of several epigenomic marks across distinct cell lines. First, we analyzed broad data from the histone modifications H3K36me3 and H3K27me3 as well as data from EZH2 (Section \ref{sec:broad}). Secondly, we assessed the performance of the presented model on ChIP-seq experiments from the transcription factor CTCF and the histone modifications H3K27ac and H3K4me3 characterized by short peaks (Section \ref{sec:sharp}). For H3K27ac and H3K4me3, enrichment peaks are usually deposited on the promoter regions of actively transcribed genes and several studies have associated their role with gene transcription \citep{creyghton2010histone,lauberth2013h3k4me3}. The transcription factor CTCF is a protein that binds to short DNA motifs and is responsible for several cellular processes that include the regulation of the chromatin 3D structure and mRNA splicing \citep{shukla2011ctcf}. Two isogenic replicates for each epigenomic mark were used in the analysis. Using RNA-seq data, we assessed the practical significance of our results by associating the detection and classification of differential combinatorial patterns from called peaks of H3K36me3, H3K27me3, and EZH2 with the direction of gene expression (Section \ref{sec:predpostprob}; see Web Appendix D1 for the RNA-seq data processing).

We compared the genome-wide performance of the presented model with the current DPCs ChIPComp, csaw, DiffBind, diffReps, RSEG, and THOR. For the methods that require a set of candidate regions to be specified \textit{a priori}, ChIPComp and DiffBind, peaks were called in advance using MACS2 (\cite{zhang2008model}; available at \url{https://github.com/macs3-project/MACS}) and used as input to the software for differential call. We benchmarked methods regarding the coverage of differentially transcribed gene bodies, the number and average size of differential peak calls, $\log_{2}$ fold change (LFC) of read counts, Spearman correlation of $\log_{2}$-transformed read counts between cell lines, and computation time. Metrics for sensitivity and specificity were defined on the window level and based on the coverage of differentially transcribed gene bodies by called peaks (Web Appendix D1). For broad marks, read counts were computed using non-overlapping windows of 500bp. For the remaining short marks, we computed read counts using non-overlapping windows of 250bp. Results presented in this section pertain to the analysis of two cell lines, namely Helas3 and Hepg2. A discussion about the choice of the window size is presented in the Web Appendix D2. Results from the analysis of more than two cell lines are presented in the Web Appendix D3. Data accessing code, data pre-processing steps, method-specific parameters, and code to replicate the presented results are detailed in the Web Appendix A.

\subsection{Analysis of ChIP-seq Data From Broad Marks}
\label{sec:broad}
Methods were benchmarked regarding the coverage of differentially transcribed gene bodies. The histone modification H3K36me3 is known to be associated with gene transcription and enriched regions of this mark are usually deposited on transcribed gene bodies. Hence, the location of differential peaks of H3K36me3 is expected to agree with the location of differentially expressed genes. Following the analysis presented by \cite{steinhauser2016comprehensive}, we defined a set of protein coding genes exhibiting $|\text{LFC}|>2$ of ChIP-seq read counts between the two analyzed cell lines as true differentially transcribed genes. Results using different threshold levels are presented in the Web Appendix D1 and agree with those presented here. Protein-coding genes with total read count across cell lines under the $25^{th}$ percentile were excluded from the analysis. Normalization by the median log-ratios of each replicate over the geometric mean was performed to avoid spurious differences due to sequencing depth.%The analysis of H3K27me3 is challenging as its enrichment is often associated with both intergenic and silenced genes. Methods that utilize a gold-standard for benchmarking purposes for this mark rely on real-time PCR results on specific loci \citep{zang2009clustering}. Here, we compare methods for this mark with respect to the characteristics of called peaks.

In Figure~\ref{fig:fig_broad}A, we show receiver operating characteristic (ROC) curves for various methods and different nominal levels of FDR threshold for differential peaks of H3K36me3. Similar to the analysis of broad histone modification marks presented by \citet{xing2012genome}, we computed the observed true positive rates (false positive rates) on the window-level as the proportion of windows called as differential out of the total number of windows associated (not associated) with differentially transcribed genes (Web Appendix D1). Our method had the best overall performance among all DPCs as its differential peaks were able to cover most of differentially transcribed gene bodies while still maintaining a low number of false positives. Methods that tended to call short peaks, such as ChIPComp and DiffBind, were the ones with the lowest sensitivity among all methods. ChIPComp and DiffBind have been previously shown to be dependent on the set of candidate peaks and to perform best in scenarios with short peaks (Figure~\ref{fig:fig_broad}B; \citet{steinhauser2016comprehensive,lun2015csaw}). In Figure~\ref{fig:fig_broad}C, we show the observed sensitivity (y-axis) and the average differential peak size (kbp; x-axis) for various methods under different nominal FDR levels (the observed FDR is annotated next to each data point). Our model and RSEG, two HMM-based methods, tended to call broader differential peaks and exhibited better sensitivity than other methods. Yet, differential peaks called by RSEG often did not correspond to differential regions of enrichment (Figure~\ref{fig:fig_broad}D), a behavior that has been noted by others \citep{starmer2016detecting} and also seen in simulated data (Figure~\ref{fig:fig_simulation_csaw}). Our HMM-based method with a non-linear normalization scheme via model offsets allowed us to maintain a low observed FDR and a higher sensitivity than other DPCs. In Figure~\ref{fig:fig_broad}F, we show examples of differential peak calls for EZH2. Our method was among the fastest algorithms due to our computational scheme, taking approximately 1 hour to analyze genome-wide data (Figure~\ref{fig:fig_broad}E).
\begin{figure}
\begin{center}
\includegraphics[scale = 0.65]{../../Main/Figure_Broad/Figure_Broad.pdf}
\end{center}
\caption{Analysis of broad ENCODE data. (A): ROC curves of H3K36me3 differential peak calls. (C): sensitivity (y-axis) and average H3K36me3 differential peak size (kbp; x-axis) of various methods under different nominal FDR thresholds (observed FDR annotated next to data points). (B), (D), and (F): example of peak calls from H3K36me3, H3K27me3, and EZH2, respectively, under a nominal FDR control of 0.05. Posterior probabilities of the HMM differential state are shown at the bottom of each panel. (E): computing time of genome-wide analysis from various methods. This figure appears in color in the electronic version of this article.
\label{fig:fig_broad}}
\end{figure}

\subsection{Analysis of ChIP-seq Data From Short Marks}
\label{sec:sharp}
We further evaluated the performance of the proposed method on data sets characterized by short peaks, namely the histone modifications H3K4me3 and H3K27ac and the transcription factor CTCF. The goal of our analysis was to assess whether our method was robust to different types of data and still able to call short differential regions of enrichment. In these scenarios, differential peaks are usually observed in isolated genomic regions and exhibit a high SNR. It has been shown that certain HMM-based approaches, including RSEG, have low accuracy under short histone modification marks and TF data \citep{hocking2016optimizing}. As we show, the model proposed in this article performs comparably to the evaluated DPCs known to perform best in short data (ChIPComp, DiffBind; \cite{steinhauser2016comprehensive}) and appeared to be more efficient regarding the computation time in certain scenarios.

We calculated the LFC and the Spearman correlation between cell lines Helas3 and Hepg2 based on ChIP-seq read counts mapped onto differential peaks called by each method. Read counts were previously normalized by the median log-ratios of each replicate over the geometric mean to avoid spurious differences due to sequencing depth. As these marks are characterized by short peaks, ideal methods would show high absolute LFC and negative correlation between read counts mapped on differential peaks. Figure~\ref{fig:fig_sharp} shows the main results from our analysis using short data sets. In Figures~\ref{fig:fig_sharp} A and C we show the median LFC and the Spearman correlation of ChIP-seq counts for differential CTCF and H3K4me3 peak calls (sorted by the absolute LFC), respectively, under a nominal FDR control of 0.05. We present separate curves regarding the signal of observed enrichment to better characterize the direction of change. The results show that the HMM-based methods RSEG and THOR were among those with the lowest absolute LFC among all methods, which confirms their sub optimal performance in the scenario of short peaks \citep{hocking2016optimizing}. In addition, we observed that ChIPComp had the best performance overall as it was able to call differential peaks with the highest absolute LFC and the lowest correlation between read counts of the two analyzed cell lines. Our model was able to properly call truly short differential peaks (Figures~\ref{fig:fig_sharp} B, D, and F) and was comparable to the non-HMM based methods regarding the computation time (Figure~\ref{fig:fig_sharp}E), jointly calling differential peaks in less than 1.5 hour.
\begin{figure}
\begin{center}
\includegraphics[scale = 0.65]{../../Main/Figure_Narrow/Figure_Narrow.pdf}
\end{center}
\caption{Analysis of short ENCODE data. (A) and (C): median LFC and correlation between cell lines of ChIP-seq counts from differential peaks for CTCF and H3K4me3, respectively. (B), (D), and (F): example of peak calls from CTCF, H3K4me3, and H3K27ac, respectively. Posterior probabilities of the HMM differential state are shown at the bottom of each panel. (E): computing time of genome-wide analysis from various methods. Results are shown under a nominal FDR control of 0.05. This figure appears in color in the electronic version of this article.
\label{fig:fig_sharp}}
\end{figure}

\subsection{Genomic Segmentation and Classification of Chromatin States}
\label{sec:predpostprob}
Lastly, we analyzed data from the cell line Helas3 to segment its genome regarding the activity of marks H3K36me3, H3K27me3, and EZH2. We considered each mark as a separate experimental condition ($G=3$) and sought to jointly classify local chromatin states in Helas3 based upon the presence or absence of enrichment from each mark. It is known that EZH2 catalyzes the methylation of H3K27me3, a repressive mark, and H3K36me3 is associated with transcribed genes (Section \ref{sec:background}). Hence, we expected regions of enrichment in consensus for these marks to be rare and differential regions to be mostly represented by either transcribed chromatin states (enrichment for H3K36me3 alone) or repressed chromatin states (enrichment co-occurrence for H3K27me3 and EZH2). The analyses presented in this section highlight the applicability of our method in the context of genomic segmentation \citep{ernst2012chromhmm}, a distinct problem not tackled by current DPCs.

First, we segmented the genome using the Viterbi sequence of most likely HMM states to understand the distribution of genomic regions associated with consensus background, differential, and consensus enrichment states (Figure~\ref{fig:fig_multimarks}). While the majority of the genomic regions exhibited no enrichment for any of the analyzed marks, regions of consensus enrichment were rare and represented only 2\% of the analyzed genome (Figure~\ref{fig:fig_multimarks}A), as expected. Consensus background and differential regions mostly covered intergenic (66\%) and protein-coding genic regions (61\%), respectively. Differential genomic windows were mostly representing either transcribed chromatin states or repressed chromatin states (Figure~\ref{fig:fig_multimarks}B). All differential combinatorial patterns expected to be rare had associated mixture proportion estimates less than 0.02 (see Web Appendix B3 for a discussion on pruning rare states). %. The Viterbi algorithm allows the simultaneous segmentation of the genome regarding the three HMM states without the necessity of choosing state-specific FDR control thresholds
\begin{figure}
\begin{center}
\includegraphics[scale = 0.7]{../../Main/Figure_MultiMarks/Figure_MultiMarks.pdf}
\end{center}
\caption{Genomic chromatin state segmentation and classification. (A): distribution of base pairs (y-axis) and the Viterbi sequence of states (x-axis). (B): estimated mixture probabilities and the associated differential combinatorial patterns. (C): density estimate from expression of genes intersecting differential peaks associated with the enrichment of H3K36me3 alone or the enrichment of H3K27me3 and EZH2 in consensus. (D): example of a genomic region with differential peaks and genes, colored according to their classification and expression levels, respectively. This figure appears in color in the electronic version of this article.
\label{fig:fig_multimarks}}
\end{figure}

These results suggest that protein-coding genes overlapping differential regions should be either silenced (e.g. genes associated with repressed chromatin states) or highly expressed (e.g. genes associated with transcribed chromatin states). To assign combinatorial patterns to differential peaks, we chose the combination pertaining to the most frequent mixture component across windows by using the maximum estimated mixture model posterior probability, $Pr(W_{jl}=1|Z_{j}=2,\mb{y}_{..j},\mb{x};\bs{\Psi}^{(t)})$, $j=1,\ldots,M$. For genes overlapping differential regions associated with either transcribed or repressed chromatin states, we computed the distribution of transcripts per million (TPM) from matching RNA-seq experiments. Genes associated with transcribed chromatin states had a significantly higher distribution of TPM counts than genes associated with repressed chromatin states (Figure~\ref{fig:fig_multimarks}C). We detected broad differential regions of enrichment and the classification of differential combinatorial patterns agreed with their biological roles as well as the expression levels of associated gene bodies. Figure~\ref{fig:fig_multimarks}D shows an example of a genomic region with differential enrichment for the analyzed marks. We compare our results to ChromHMM with 3 states, a method developed for chromatin segmentation. Our method offers the benefit of simultaneously detecting differential peaks and classifying the combinatorial pattern of enrichment through mixture model posterior probabilities even in the context of genomic segmentation with highly diverse epigenomic marks (see Web Appendix D4). By using the BIC for model selection (Web Appendix B3), one can choose the number of biologically relevant mixture components to be included in the model, a task that may not be as straightforward in methods such as ChromHMM (see Supplementary Figure 4 in \citet{ernst2012chromhmm}).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Discussion}
\label{sec:discussion}

We presented a flexible and efficient statistical model designed to call differential regions of enrichment from ChIP-seq experiments with multiple replicates and multiple conditions. Our model has three main advantages over current methods tailored for differential peak detection. First, it uses an HMM-based approach that accounts for the local dependency of ChIP-seq counts and is able to precisely detect broad and short differential regions of enrichment. Second, it utilizes a GLM-based framework with model offsets that account for potential non-linear biases in the data as well as a constrained parametrization across HMM states and mixture components. Our implementation of the RCEM algorithm led to genome-wide analyses of ChIP-seq data under a computational time comparable to some of the fastest current methods and was at least 5 times faster than current HMM-based algorithms. Lastly, our method allows the simultaneous detection and classification of differential combinatorial patterns of enrichment from its embedded mixture model and the associated posterior probabilities under any number of conditions. Our software has been implemented into an R package and is available for download (Web Appendix A). % Data resulting from these assays are similar in nature but aim to identify regions that alternatively are not bound to proteins. % Such a tool could trigger, for instance, new insights on the field of epigenomics as investigators are often interested in making sense of the roles of epigenomic marks on the human genome. % Analysis of datasets from ther epigenomic assays, such as ATAC-seq, could also benefit from the presented method.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\backmatter
% \section*{Acknowledgements}
% The authors thank Professor A. Sen for some helpful suggestions,
% Dr C. R. Rangarajan for a critical reading of the original version of the
% paper, and an anonymous referee for very useful comments that improved
% the presentation of the paper.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\bibliographystyle{biom}\bibliography{bibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Supporting Information}
Web Appendices A-D, referenced in Sections 1-6, and a link to the implemented software
are available with this article at the Biometrics website on Wiley Online Library.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\label{lastpage}
\end{document}

