%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\documentclass[useAMS,usenatbib,referee]{biom}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newcommand{\bs}[1]{\boldsymbol{#1}}
\newcommand{\mb}[1]{\mathbf{#1}}
\newcommand*\red{\color{red}}

\usepackage[figuresright]{rotating}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{enumerate}
\usepackage{natbib}
\usepackage{url}
\usepackage{graphicx}
\usepackage{subfig}
\usepackage{booktabs}
%\usepackage{lipsum}
\usepackage{array}
\usepackage{float}
\usepackage{multirow}
\usepackage{appendix}
\usepackage{makecell}
\usepackage{color}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\setcounter{footnote}{2}

\title{Efficient Detection and Classification of Epigenomic Changes Under Multiple Conditions}

\author{Pedro L. Baldoni$^*$\email{baldoni@email.unc.edu},
Naim U. Rashid$^{**}$\email{naim@unc.edu}, and
Joseph G. Ibrahim$^{***}$\email{ibrahim@bios.unc.edu} \\
Department of Biostatistics, University of North Carolina at Chapel Hill, 
Chapel Hill, North Carolina, USA}

\begin{document}

\date{{\it Received May} 2020. {\it Revised May} 2020. {\it Accepted May} 2020.}

\pagerange{\pageref{firstpage}--\pageref{lastpage}}\pubyear{2020}

\volume{1}
\artmonth{May}
\doi{00.0000/0.000.0.0.0}

\label{firstpage}

\begin{abstract}
Epigenomics, the study of the human genome and its interactions with proteins and other cellular elements, has become of significant interest in recent years. Such interactions have been shown to regulate essential cellular functions and are associated with multiple complex diseases. Therefore, understanding how these interactions may change across conditions is central in biomedical research. Chromatin immunoprecipitation followed by massively-parallel sequencing (ChIP-seq) is one of several techniques to detect local changes in epigenomic activity (peaks). However, existing methods for differential peak calling are not optimized for the diversity in ChIP-seq signal profiles, are limited to the analysis of two conditions, or cannot classify specific patterns of differential change when multiple patterns exist. To address these limitations, we present a flexible and efficient method for the detection of differential epigenomic activity across multiple conditions. We utilize data from the ENCODE Consortium and show that the presented method, epigraHMM, exhibits superior performance to current tools and it is among the fastest algorithms available, while allowing the classification of combinatorial patterns of differential epigenomic activity and the characterization of chromatin regulatory states.
\end{abstract}

\begin{keywords}
%  Please place your key words in alphabetical order, separated
%  by semicolons, with the first letter of the first word capitalized,
%  and a period at the end of the list.
ChIP-seq; Differential Peak Call; Epigenomics; Hidden Markov Model; Mixture Model.
\end{keywords}

\maketitle

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}
\label{sec:introduction}

% Paragraph 1: Why does epigenetics matter? Nail down the problem 
Epigenomics, the study of the genome and its interactions with proteins and cellular elements, has become of significant interest in recent years. Such interactions may regulate essential cellular functions, such as gene expression, resulting in downstream phenotypic impact \citep{kim2018polycomb}. Hence, the interrogation of how these interactions may change across conditions, such as cell types or treatments, is of marked interest in biomedical research. Several landmark articles have identified genomic regions of changing (differential) epigenomic activity between conditions as drivers of cell differentiation \citep{creyghton2010histone} and a number of human diseases \citep{portela2010epigenetic}. Within differential regions, delineating specific patterns of change across conditions is also of interest, for example classifying the gain-of- or loss-of-activity in genomic loci due to treatment \citep{clouaire2014cfp1}. %The identification of specific combinations of processes acting locally may also be informative, such as for segmenting the genome into regulatory states \citep{kundaje2015integrative}.

% Paragraph 3: Introduce ChIP-seq
To quantify local epigenomic activity, a common high-throughput assay is chromatin immunoprecipitation followed by massively parallel sequencing (ChIP-seq). ChIP-seq begins with cross-linking DNA and proteins within chromatin structures, which are then fragmented by sonication. DNA fragments bound to the protein of interest are isolated by immunoprecipitation and sequenced via high-throughput sequencing to generate short reads pertaining to the original fragments. Sequences are then mapped onto a reference genome to determine their likely locations of origin. Genomic coordinates containing a high density of mapped reads (enrichment regions, peaks) indicate likely locations of protein-DNA interaction sites, while other regions are referred to as background regions. The read density is often summarized by counting reads mapped onto non-overlapping windows of fixed length (window read counts), forming the basis for downstream analyses. Across multiple conditions, regions exhibiting enrichment in at least one condition, but not across all conditions, indicate the presence of differential activity pertaining to the protein-DNA interaction of interest.

% Paragraph 4: Issues with the data/methods
To date, many differential peak callers (DPCs) have been proposed. However, several challenges affect their ability to accurately detect regions of differential activity from the wide range of experiments (Section \ref{sec:background}). First, differential regions may be both short or broad, causing difficulty for methods optimized for a particular type of peak profile \citep{stark2011diffbind,chen2015novel}. Second, pooling experimental replicates often leads to lower specificity in comparison to joint modeling samples from each condition \citep{song2011identifying}. Third, ChIP-seq data is often subject to complex biases that vary across the genome, as differences in local read enrichment may depend on quantities such as local read abundance or GC content \citep{teng2017accounting}. DPCs that ignore such effects, or rely on global scaling factors or control subtraction methods \citep{shen2013diffreps,allhoff2016differential}, may be subject to spurious differences due to the lack of non-linear normalization methods \citep{lun2015csaw}. In ChIP-seq data with broad peaks, a recent comparison showed that methods often detect a large number of either short or false positive peaks \citep{steinhauser2016comprehensive}. Moreover, only few methods can detect or classify differential combinatorial patterns across any number of conditions \citep{stark2011diffbind,chen2015novel,lun2015csaw}.
 
% Paragraph 4: What we propose? Broad peaks, mixture model, and multiple conditions
Here, we present epigraHMM (Figure~\ref{fig:fig_motivation}), an efficient and flexible statistical method to identify differential regions of enrichment from epigenomic experiments with diverse signal profiles. We overcome the limitations of current DPCs with three major features. First, it uses a hidden Markov model (HMM) to account for the diversity in differential enrichment from broad and short ChIP-seq data collected under multi-replicate, multi-condition settings. Second, it models specific combinatorial patterns of enrichment via a finite mixture model emission distribution within the HMM differential state. Each mixture component pertains to a particular differential combinatorial pattern originated by the presence or absence of local enrichment across conditions, where a generalized linear model (GLM) encodes specific differential combinatorial patterns while accounting for sample- and window-specific normalizing offsets. Third, it simultaneously detects and classifies epigenomic changes under three or more conditions, a novelty not yet available in any other DPC algorithm. epigraHMM offers benefits over current HMM algorithms \citep{song2011identifying,allhoff2016differential} with an embedded GLM mixture model that allows the modeling of covariates of interest, the inclusion of normalizing offsets for non-linear biases (such as GC-content bias), and a fast estimation scheme via rejection-controlled EM algorithm (RCEM; \cite{ma2006data}).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Data}
\label{sec:background}

% What are histones and why are they important
Histones are proteins that condense the DNA in eukaryotic cells into units called nucleosomes and, when enzymatically modified, mediate changes in local DNA packaging and chromatin accessibility. Hence, cellular processes such as gene transcription, gene silencing, and DNA repair are also affected. Proteins that interact with DNA and alter its functional properties are often referred to as epigenomic marks. The histone modifications H3K36me3 and H3K27me3 are example of marks that associate with genomic loci containing transcribed and repressed genes \citep{liu2016distinct}, respectively, while exhibiting broad enrichment profiles. The enhancer of zeste homolog 2 (EZH2), a component of the complex PRC2 that catalyzes the methylation of H3K27me3 \citep{margueron2011polycomb}, is another example of a protein characterized by broad enrichment domains and co-occurs with the activity of H3K27me3. Conversely, short peaks from histone modifications H3K27ac and H3K4me3 are usually deposited on the promoter regions of transcribed genes and several studies have associated their role with gene transcription \citep{creyghton2010histone}. Similarly, the transcription factor CTCF is a protein that binds to short DNA motifs and is responsible for cellular processes that include the regulation of the chromatin 3D structure \citep{shukla2011ctcf}.

Using ChIP-seq data pertaining to histone modifications H3K27me3 and H3K36me3, and to EZH2 from the ENCODE Consortium \citep{landt2012chip}, we find that current DPCs have difficulty in accurately detecting broad regions of differential enrichment between several common cell lines (Figure~\ref{fig:fig_motivation}). In line with previous findings \citep{steinhauser2016comprehensive}, we observe that even current DPCs designed for broad data \citep{song2011identifying,allhoff2016differential} tend to either call overly fragmented differential peaks or call regions exhibiting no difference in experimental signal between conditions as differential (Sections \ref{sec:sim2} and \ref{sec:broad}). Methods that rely on candidate peaks may also exhibit a compromised performance due to the limitations of single-sample peak callers in broad data \citep{stark2011diffbind,chen2015novel}. Moreover, most DPCs restrict their application to the analysis of two experimental conditions. For DPCs tailored for the analysis of three or more conditions, the classification of specific differential combinatorial patterns across conditions (or across various epigenomic processes) is still an open problem. The classification of such patterns would allow researchers to quantify treatment responses on the epigenomic level or identify sets of processes working together to regulate local chromatin state \citep{clouaire2014cfp1}.

\begin{figure}
\begin{center}
\includegraphics[scale = 0.75]{../../Main/Figure_epigraHMM/Figure_epigraHMM.pdf}
\end{center}
\caption{At the top, differential peak calls from current methods (under false discovery rate control of 0.05) for the histone modification H3K27me3 among cell lines Helas3, Hepg2, and Huvec. Shaded regions indicate  observed differential enrichment, and each vertical line type bordering each region represents a different combinatorial pattern of enrichment across cell lines. Optimal methods would call broad peaks inside shaded regions and no peaks outside them. At the bottom, general overview of the presented method for simultaneous detection and classification of broad and short differential regions of enrichment. This figure appears in color in the electronic version of this article.
\label{fig:fig_motivation}}
\end{figure}

We assessed the performance of our model on ChIP-seq experiments characterized by broad peaks (H3K36me3, H3K27me3, and EZH2) and short peaks (H3K27ac, H3K4me3, and CTCF). In simulated and in real data from the ENCODE Consortium (Sections \ref{sec:simulation} and \ref{sec:realdata}), our model addresses the issues of the current peak callers in broad data, while being flexible for short peaks and comparable to the fastest DPCs regarding the computation time. We show that our method can also be utilized for genomic regulatory state segmentation when studying multiple types of epigenomic processes from a single condition or cell line (Section \ref{sec:predpostprob}). Web Appendices A1-A4 present data accession codes, data pre-processing steps, code implementing the method, and scripts to replicate the presented results, respectively.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Methods}
\label{sec:methods}

\subsection{Statistical Model}
\label{subsec:hmmmixt}
Let $Y_{hij}$ denote the random variable pertaining to the read count for genomic window $j$ from sample $i$ of condition $h$, where $j=1,\ldots,M$, $i=1,\ldots,n_{h}$, $h=1,\ldots,G$, and let $y_{hij}$ be the observed count. Here, $n_h$ is the number of samples in condition $h$ and $N = \sum_{h = 1}^G n_h$ is the total number of samples across the $G$ conditions.  At the $j^{th}$ window, let $\mb{y}_{..j}=(y_{11j},\ldots,y_{Gn_{G}j})^\prime$ denote the $N \times 1$ vector of window read counts across all samples and conditions, and let $\mb{y}=(\mb{y}_{..1},\ldots,\mb{y}_{..M})$ denote the corresponding $N \times M$ matrix of window read counts spanning all windows, samples, and conditions. We assume that each window belongs to one of three possible hidden states: consensus background, differential, and consensus enrichment. Windows exhibiting low (high) enrichment across all conditions will be modeled by an emission distribution pertaining to the consensus background (enrichment) state. Windows exhibiting enrichment under at least one condition, but not all conditions, will be modeled by an emission distribution pertaining to the differential state. If $G$ conditions are of interest, there are $L = 2^G-2$ possible differential combinatorial patterns of enrichment and background across conditions at a given window. The emission distribution pertaining to the differential state models all $L$ possible differential combinatorial patterns via a mixture model with mixture proportions $\bs{\delta} = (\delta_{1},\ldots,\delta_{L})^\prime$, such that $\sum_{l=1}^{L}\delta_{l}=1$.

We assume a single latent discrete time stationary Markov chain $\mb{Z}=\{Z_{j}\}_{j=1}^{M}$, $Z_{j}\in\{1,2,3\}$, with state-to-state transition probabilities $\bs{\gamma}=(\gamma_{11},\gamma_{12},\ldots,\gamma_{33})^\prime$ and initial probabilities $\bs{\pi}=(\pi_{1},\pi_{2},\pi_{3})^\prime$, such that $\sum_{s=1}^{3}\gamma_{rs}=1$ and $\sum_{s=1}^{3}\pi_{s}=1$ for $r\in\{1,2,3\}$.  Let $f_{r}(\mb{y}_{..j}| \bs{\psi}_{r})$ denote the emission distribution corresponding to the $r^{th}$ hidden state, $\bs{\Psi} = (\bs{\pi}^\prime,\bs{\gamma}^\prime,\bs{\delta}^\prime,\bs{\psi}^\prime)^\prime$ denote the vector of all model parameters, $\bs{\psi}=(\bs{\psi}_{1}^\prime,\bs{\psi}_{2}^\prime,\bs{\psi}_{3}^\prime)^\prime$ denote each state's set of emission distribution-specific parameters, and $\mathcal{Z}$ denote the set of $3^M$ possible state paths of length $M$. The likelihood function of the proposed HMM may be written as

\begin{align}\label{eq:difflik}
f(\mathbf{y}|\mathbf{x};\bs{\Psi}) &= \sum_{\mb{Z}\in\mathcal{Z}}\left\{\prod_{r=1}^{3}\pi_{r}^{I(Z_{1}=r)}\times\left(\prod_{j=2}^{M}\prod_{r=1}^{3}\prod_{s=1}^{3}\gamma_{rs}^{I(Z_{j-1}=r,Z_{j}=s)}\right)\right.\times\\
&\times\left.\left(\prod_{j=1}^{M}f_{1}\left(\mb{y}_{..j}|\bs{\psi}_{1}\right)^{I(Z_{j}=1)}f_{2}\left(\mb{y}_{..j}|\mb{x};\bs{\delta},\bs{\psi}_{2}\right)^{I(Z_{j}=2)}f_{3}\left(\mb{y}_{..j}|\bs{\psi}_{3}\right)^{I(Z_{j}=3)}\right)\right\}.\nonumber
\end{align}

Here, $\mb{x}$ is a fixed $G \times L$ design matrix enumerating each of the $L$ possible differential combinatorial patterns in terms of the presence or absence of enrichment across each of the $G$ conditions, only in the emission distribution of the differential state. We assume that read counts pertaining to genomic windows from the consensus background ($r=1$) and consensus enrichment  ($r=3$) states follow a Negative Binomial (NB) distribution with state-specific parameters $\bs{\psi}_{r}=(\mu_{(r,hij)},\phi_{r})^\prime$, with mean $\mu_{(r,hij)}$ and variance $\mu_{(r,hij)}(1+\mu_{(r,hij)}/\phi_{r})$. Conditional on the HMM state, we assume independence of read counts across samples and write the emission distribution of the consensus background and enrichment states as
\begin{align}\label{eq:nondiffdensity}
f_{r}(\mb{y}_{..j}|\bs{\psi}_{r}) &= \prod^{G}_{h=1}\prod^{n_{h}}_{i=1}\frac{\Gamma(y_{hij}+\phi_{r})}{y_{hij}!\Gamma(\phi_{r})}\left(\frac{\phi_{r}}{\mu_{(r,hij)}+\phi_{r}}\right)^{\phi_{r}}\left(\frac{\mu_{(r,hij)}}{\mu_{(r,hij)}+\phi_{r}}\right)^{y_{hij}},\quad r\in\{1,3\},
\end{align}
such that $\log(\mu_{(1,hij)})=\beta_{1} + u_{hij}$, $\log(\phi_{1})=\lambda_{1}$, $\log(\mu_{(3,hij)})=\beta_{1} + \beta_{3} + u_{hij}$, and $\log(\phi_{3})=\lambda_{1} + \lambda_{3}$. The offset $u_{hij}$ may adjust for technical or biological artifacts such as the GC-content bias and allows the non-linear normalization of read counts across genomic windows, replicates, and conditions (Web Appendix B1; \cite{teng2017accounting}). %When $u_{hij}=0$, $\beta_{1}$ and $\lambda_{1}$ represent the log-mean and log-dispersion, respectively, of read counts pertaining to consensus background state windows, whereas $\beta_{3}$ and $\lambda_{3}$ represent the difference in log-mean and log-dispersion of read counts from consensus enrichment state windows relative to consensus background state windows.

In the differential state ($r = 2$), read counts are modeled by a $L$-component finite mixture model with mixture components that follow a NB distribution, where each component corresponds to a particular differential combinatorial pattern. To define these patterns, consider sets $S_{1},\ldots,S_{L}$  delineating the subset of the $G$ conditions that are enriched in each of the $L$ differential combinatorial patterns. For instance, if $G=3$, the sets $S_{1} =\{1\}$, $S_{2} =\{2\}$, $S_{3} =\{3\}$, $S_{4} =\{1,2\}$, $S_{5} =\{1,3\}$, and $S_{6} =\{2,3\}$ define the six possible differential combinatorial patterns of enrichment and background across three conditions (\textit{e.g.}, $S_{6}$ denotes enrichment in conditions 2 and 3 and background in condition 1). The presence or absence of enrichment in each of the $L$ sets is encoded into each column of $\mb{x}=(\mb{x}_{1},\ldots,\mb{x}_{L})$, such that $\mb{x}_{l} = (x_{1l},\ldots,x_{Gl})'$, and $x_{hl} = I(h\in S_{l})$ for $l=1,\ldots,L$ and $h = 1,\ldots,G$. That is, $\mb{x}_{l}$ is the $G \times 1$ vector of binary indicator variables denoting which subset of conditions are enriched in pattern (mixture component) $l$. Let $\bs{\psi}_{2}$ and $\bs{\psi}_{(2,l)}$ denote the parameter vectors pertaining to the differential state and to the $l^{th}$ mixture component, respectively. Conditional on the differential state, we assume independence of read counts across samples and write the finite mixture model emission distribution as
\begin{align}\label{eq:diffdensity}
\resizebox{.925\hsize}{!}{$f_{2}(\mb{y}_{..j}|\mb{x};\bs{\delta},\bs{\psi}_{2}) = \sum_{l=1}^{L}\delta_{l}\prod_{h=1}^{G}\prod_{i=1}^{n_{h}}\frac{\Gamma\left(y_{hij}+\phi_{(2,lh)}\right)}{y_{hij}!\Gamma\left(\phi_{(2,lh)}\right)}\left(\frac{\phi_{(2,lh)}}{\mu_{(2,lhij)}+\phi_{(2,lh)}}\right)^{\phi_{(2,lh)}}\left(\frac{\mu_{(2,lhij)}}{\mu_{(2,lhij)}+\phi_{(2,lh)}}\right)^{y_{hij}},$}
\end{align}
where $\mu_{(2,lhij)}$ and $\phi_{(2,lh)}$ are the mean and dispersion, respectively, pertaining to read counts originating from window $j$ and sample $i$ in condition $h$ from the mixture component $l$. We assume that $\log(\mu_{(2,lhij)}) = \beta_{1} + \beta_{3}x_{hl} + u_{hij}$ and $\log(\phi_{(2,lh)}) = \lambda_{1} + \lambda_{3}x_{hl}$. That is, in the mixture component $l$, we utilize the same consensus background (consensus enriched) parametrization from (\ref{eq:nondiffdensity}) in all conditions that $\mb{x}_{l}$ specifies to be background (enriched) in the $l^{th}$ differential combinatorial pattern. Such a parametrization ensures that windows exhibiting differential enrichment across conditions share common means and dispersions between the consensus background enrichment states, an assumption that increases computational efficiency. Utilizing a mixture model as the differential state emission distribution avoids the computational burden that would come from assuming separate hidden states for each of the $L$ differential combinatorial patterns.  We evaluate the strength of these assumptions through multiple simulations and a real data benchmarking analysis in Sections \ref{sec:simulation} and \ref{sec:realdata}.

\subsection{Estimation}
\label{subsec:hmmest}
Here, consider a set of latent variables $\mb{W}=(\mb{W}_{1}^\prime,\ldots,\mb{W}_{M}^\prime)^\prime$, such that $\mb{W}_{j}=(W_{j1},\ldots,W_{jL})^\prime$ for $j=1,\ldots,M$. We assume that $\mb{W}$ is a sequence of independent random vectors such that $\mb{W}_{j}|(Z_{j}=2)\sim\text{Multinomial}(1,\bs{\delta})$ and $\mb{W}_{j}|(Z_{j}=r)=0$ with probability 1 if $r=\{1,3\}$. The data generating mechanism can be then interpreted as read counts pertaining to window $j$ being sampled from $f_{(2,l)}$, given $\bs{\psi}_{(2,l)}$ and $\mb{x}_{l}$, when $Z_{j}=2$ (differential state) and $W_{jl}=1$ ($l^{th}$ differential combinatorial pattern). Denoting $\mathcal{W}$ as the set of $L^M$ possible combinations of latent vectors $\mb{W}$, the likelihood function of the observed data (\ref{eq:difflik}) can be rewritten as

\begin{align}\label{eq:obslik}
f(\mathbf{y}|\mathbf{x};\bs{\Psi}) &= \sum_{\mb{Z}\in\mathcal{Z}}\sum_{\mb{W}\in\mathcal{W}}\left\{\left[\prod_{r=1}^{3}\pi_{r}^{I(Z_{1}=r)}\prod_{j=2}^{M}\prod_{r=1}^{3}\prod_{s=1}^{3}\gamma_{rs}^{I(Z_{j-1}=r,Z_{j}=s)}\right.\right]\prod_{j=1}^{M}\prod_{l=1}^{L}\delta_{l}^{W_{jl}I(Z_{j}=2)}\times\\
&\times \left.\left[\prod_{j=1}^{M}f_{1}\left(\mb{y}_{..j}|\bs{\psi}_{1}\right)^{I(Z_{j}=1)}\left(\prod_{l=1}^{L}f_{(2,l)}(\mb{y}_{..j}|\mb{x}_l;\bs{\psi}_{(2,l)})^{W_{jl}}\right)^{I(Z_{j}=2)}f_{3}\left(\mb{y}_{..j}|\bs{\psi}_{3}\right)^{I(Z_{j}=3)}\right]\right\},\nonumber %&= \sum_{\mb{Z}\in\mathcal{Z}}\left\{\left[\sum_{\mb{W}\in\mathcal{W}}f(\mb{y}|\mb{w},\mb{z},\mb{x};\bs{\psi})f(\mb{w}|\mb{z},\mb{x};\bs{\delta})\right]f(\mb{z}|\mb{x};\bs{\pi},\bs{\gamma})\right\},\\
\end{align}
where $f_{(2,l)}(\mb{y}_{..j}|\mb{x}_l;\bs{\psi}_{(2,l)})$ is defined as in (\ref{eq:diffdensity}). In the $t^{th}$ iteration of the EM algorithm, the $Q$ function of the complete data log-likelihood can be written as
\begin{align}\label{eq:diffQfunc}
Q\left(\bs{\Psi}|\bs{\Psi}^{(t)}\right) &= E_{\mb{Z}}\left(E_{\mb{W}|\mb{Z}}\left(\log\left(f(\mb{y},\mb{W},\mb{Z}|\mb{x};\bs{\Psi})\right)|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right)|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right),\nonumber\\
&= Q_{0}(\bs{\pi},\bs{\gamma}|\bs{\Psi}^{(t)})+Q_{1}(\bs{\psi}_{1}|\bs{\Psi}^{(t)})+Q_{2}(\bs{\delta},\bs{\psi}_{2}|\bs{\Psi}^{(t)})+Q_{3}(\bs{\psi}_{3}|\bs{\Psi}^{(t)}),
\end{align}
where $Q_{0}(\bs{\pi},\bs{\gamma}|\bs{\Psi}^{(t)})$, $Q_{1}(\bs{\psi}_{1}|\bs{\Psi}^{(t)})$, $Q_{2}(\bs{\delta},\bs{\psi}_{2}|\bs{\Psi}^{(t)})$, and $Q_{3}(\bs{\psi}_{3}|\bs{\Psi}^{(t)})$ are defined in Web Appendix B4. In the E-step of the EM algorithm, we compute the posterior probabilities from (\ref{eq:diffQfunc}). The quantities $Pr\left(Z_{j}=r|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right)$ and $Pr\left(Z_{j-1}=r,Z_{j}=s|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right)$ can be calculated through the Forward-Backward algorithm (Web Appendix B4) and $Pr(W_{jl}=1|Z_{j}=2,\mb{y}_{..j},\mb{x};\bs{\Psi}^{(t)})=f_{(2,l)}(\mb{y}_{..j}|\mb{x}_l;\bs{\psi}_{(2,l)}^{(t)})\delta_{l}^{(t)}/\sum_{k=1}^{L}f_{(2,k)}(\mb{y}_{..j}|\mb{x}_k;\bs{\psi}_{(2,k)}^{(t)})\delta_{k}^{(t)}$ for $l=1,\ldots,L$.

The $Q$ function is maximized with respect to the parameters $\bs{\Psi}=(\bs{\pi}^\prime,\bs{\gamma}^\prime,\bs{\delta}^\prime,\beta_{1},\beta_{3},\lambda_{1},\lambda_{3})^\prime$ during the M-step of the algorithm. Estimates of the initial and transition probabilities can be directly calculated as $\hat{\pi}^{(t+1)}_{r}=Pr\left(Z_{1}=r|\mb{y},\mb{x};\bs{\Psi}^{(t)}\right)$ and $\hat{\gamma}^{(t+1)}_{rs}=\sum_{j=2}^{M}Pr(Z_{j-1}=r,Z_{j}=s|\mb{y},\mb{x};\bs{\Psi}^{(t)})/\sum_{j=2}^{M}Pr(Z_{j-1}=r|\mb{y},\mb{x};\bs{\Psi}^{(t)})$, respectively, restricted to $\sum_{r=1}^{3}\hat{\pi}^{(t+1)}_{r}=1$ and $\sum_{s=1}^{3}\hat{\gamma}^{(t+1)}_{rs} = 1$, for $r\in\{1,2,3\}$. We perform conditional maximizations to estimate the remaining parameters $(\bs{\delta}^\prime,\beta_{1},\beta_{3},\lambda_{1},\lambda_{3})^\prime$. First, mixture proportions can be estimated as $\hat{\delta}_{l}^{(t+1)}=\sum_{j=1}^{M}Pr(Z_{j}=2|\mb{y},\mb{x};\bs{\Psi}^{(t)})Pr(W_{jl}=1|Z_{j}=2,\mb{y}_{..j},\mb{x};\bs{\Psi}^{(t)})/\sum_{j=1}^{M}Pr(Z_{j}=2|\mb{y},\mb{x};\bs{\Psi}^{(t)})$. Then, estimates of $(\beta_{1},\beta_{3},\lambda_{1},\lambda_{3})^\prime$ are obtained in a similar fashion to the parameter estimation of a weighted NB regression model (Web Appendix B4).

The EM algorithm iterates until the maximum absolute relative change in the parameter estimates three iterations apart is less than $10^{-3}$ for three consecutive iterations. Here, we use a RCEM algorithm with threshold 0.05, which substantially reduces the dimensionality of the data during the M-step by randomly assigning a zero posterior probability to genomic windows unlikely to belong to each of the HMM states. Moreover, our estimation scheme allows genomic windows with equal distribution of counts across replicates and conditions to have their posterior probability aggregated during the M-step of the algorithm, which leads to a fast gradient-based optimization. Upon convergence, HMM posterior probabilities can be used to segment the genome into consensus background, differential, or consensus enrichment windows. Approaches that control the total false discovery rate (FDR) via posterior probabilities \citep{efron2001empirical} or that estimate the most likely sequence of hidden states can be used. Let $\hat{\rho}_{j2} = Pr\left(Z_{j}=2|\mb{y},\mb{x};\hat{\bs{\Psi}}\right)$ denote the estimated posterior probability that the $j^{th}$ genomic window belongs to the differential HMM state, $j=1,\ldots,M$. A posterior probability cutoff $\alpha$ is then chosen by controlling the total FDR $\sum_{j=1}^{M}(1-\hat{\rho}_{j2})I(\hat{\rho}_{j2}\geq 1-\alpha)/\sum_{j=1}^{M}I(\hat{\rho}_{j2}\geq 1-\alpha)$, where $I(\cdot)$ is an indicator function. Differential peaks are formed by merging adjacent windows that either meet a nominal FDR level for the differential HMM state or belong to the same predicted state (Web Appendix B3). Under this set up, we are able not only to detect differential enrichment regions across multiple conditions, but we can also classify various differential combinatorial patterns of enrichment within broad and short domains with mixture model posterior probabilities.

The estimation scheme is robust to situations where certain differential combinatorial patterns of enrichment are rare (Figure \ref{fig:fig_multimarks}), which results from the fact that ChIP-seq experiments often provide enough data to estimate the parameters $(\beta_{1},\beta_{3},\lambda_{1},\lambda_{3})^\prime$ shared across all $L$ mixture components and HMM states. If pruning differential combinatorial patterns of the mixture model is of interest, the optimal number of mixture components $L^{*}$, $L^{*}<L$, can be selected via the Bayesian Information Criterion (BIC) for HMMs. We observed that selecting the optimal $L^{*}$ and combinatorial patterns via BIC agrees with pruning of rare differential combinatorial patterns that one would not expect to observe biologically (Web Appendix B2). In addition, the current implementation of epigraHMM allows for the adjustment for external covariates, such as GC-content and input controls, via model offsets (Web Appendix A3). In real data analyses, however, we did not observe a significant influence of input controls on the sensitivity and specificity of differential peak calls, a fact that has also been noted by others \citep{lun2015csaw}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Simulation Studies}
\label{sec:simulation}

We evaluated epigraHMM in two simulation studies. First, we simulated read count-based data to assess the precision of the estimation scheme, the performance of differential peak detection, and the accuracy of the classification of differential combinatorial patterns (Section~\ref{sec:sim1}). Next, we utilized the simulation pipeline from \cite{lun2015csaw} to simulate ChIP-seq reads from experiments with broad differential peaks (Section~\ref{sec:sim2}). The aim of the second simulation study was to compare epigraHMM with other DPCs in a more realistic scenario with broad peaks, while avoiding the choice of a parametric model for the data.

\subsection{Read Count Simulation}
\label{sec:sim1}
Read counts were simulated under scenarios that varied regarding the type of histone modification mark (H3K36me3 and H3K27me3), number of windows ($M$, $10^5$, $5\times 10^5$, and $10^6$ windows), number of conditions ($G$, 2, 3, and 4), and number of replicates per condition ($n$, 1, 2, and 4). We further assessed epigraHMM under different Signal-to-Noise Ratio (SNR) levels, here defined as the ratio between the means of consensus enrichment and background emission distributions, by decreasing the mean ratio while maintaining the mean-variance relationship. Model parameters were estimated from ENCODE data (Web Appendix C1). Simulated counts followed a NB distribution and were generated using a first-order Markov chain with $2^G$ states, representing every combination of background and enrichment across $G$ conditions. We assessed whether epigraHMM was able to assign all $2^G-2$ differential states to the differential HMM state, while precisely estimating model parameters and accurately classifying differential combinatorial patterns.

Table~\ref{Table:table_simulation} presents results relative to the H3K27me3 simulation scenario with $10^6$ genomic windows and ENCODE-estimated SNR (mean ratio of 3.21; see Web Appendix C1 for additional details). Depending on the number of conditions, the relative bias and the range of the reported percentiles tended to decrease as more replicates were included in the analyses. This effect was more pronounced when estimating $\beta_{3}$ and $\lambda_{3}$ in scenarios with four conditions, which highlights the importance of experimental replicates to achieve precise parameter estimates. Overall, the proposed estimation scheme led to precise parameter estimates and was robust to a data generating mechanism that was different than the one assumed by the proposed model. No significant differences regarding the relative bias of parameter estimates were observed across simulations under different number of genomic windows.

\begin{center}
\input{../../Main/Table_Simulation_ReadCounts/Table_Simulation_ReadCounts.tex}
\end{center}

\subsection{Sequencing Read Simulation}
\label{sec:sim2}
Here, we compared epigraHMM with the widely used DPCs ChIPComp, csaw, DiffBind, diffReps, RSEG, and THOR, a list that covers a variety of algorithms for both broad and short ChIP-seq enrichment profiles, on data simulated with the pipeline from \cite{lun2015csaw} without a particular model assumption. Sequencing reads from ChIP-seq experiments were generated for two conditions and two replicates per condition. For the two-step DPCs ChIPComp and DiffBind, we followed \cite{lun2015csaw} and called peaks in advance using HOMER \citep{heinz2010simple}, which were then used as input in the respective software for differential call. A hundred simulated data sets were generated and peaks were called under multiple nominal FDR thresholds. For epigraHMM and RSEG, window-based posterior probabilities were used to control the total FDR (Section \ref{subsec:hmmest}).

Overall, epigraHMM showed the highest observed sensitivity among all DPCs while maintaining an observed FDR close to the nominal value across a wide range of nominal FDR levels (Figure~\ref{fig:fig_simulation_csaw}A). Here, the observed FDR is the proportion of windows incorrectly called as differential out of the total number of windows called as differential. Due to the excessive number of false positive differential peak calls, diffReps, RSEG, and THOR showed higher observed FDR levels than the nominal FDR values (Figure~\ref{fig:fig_simulation_csaw}D). Overall, diffReps and THOR called an excessive number of short and discontiguous peaks while RSEG called regions that were usually wider than the simulated differential peaks and did not correspond to simulated differential enrichment (Figure~\ref{fig:fig_simulation_csaw}B). Regarding the computation time, epigraHMM was among the fastest DPCs due to the efficient RCEM implementation (Figure~\ref{fig:fig_simulation_csaw}C). Other HMM-based algorithms, RSEG and THOR, appeared to be the most computationally intensive and required longer amounts of time to analyze the data. In general, epigraHMM was able to consistently cover most of true differential regions with broad peaks while exhibiting a limited number of false discoveries (Figure~\ref{fig:fig_simulation_csaw}D; Web Appendix C2). Results shown for RSEG do not include 27 cases in which the algorithm failed to analyze the data due to internal errors or called the entire genome as differential. Similar issues of RSEG have been reported by others \citep{starmer2016detecting}.

\begin{figure}
\begin{center}
\includegraphics[scale = 0.65]{../../Main/Figure_Simulation_SequencingReads/Figure_Simulation_SequencingReads.pdf}
\end{center}
\caption{Sequencing read-based simulation from the csaw pipeline. (A): average observed sensitivity and FDR for various methods. (B): scatter plot of average ratio of called and simulated peaks (y-axis) and number of called peaks intersecting true differential regions (x-axis). (C): box plot of computing time (in minutes) for various algorithms. (D): an example of differential peak calls under a nominal FDR control of 0.05. Shaded areas indicate true differential peaks. This figure appears in color in the electronic version of this article.
\label{fig:fig_simulation_csaw}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Application to ENCODE Data}
\label{sec:realdata}

We applied epigraHMM on ChIP-seq data from the ENCODE Consortium to detect differential peaks of several epigenomic marks across multiple cell lines. We analyzed data characterized by broad peaks (H3K36me3, H3K27me3, and EZH2; Section \ref{sec:broad}) and short peaks (CTCF, H3K27ac, and H3K4me3; Section \ref{sec:sharp}). Two isogenic replicates of each cell line were used in the analysis. Using RNA-seq data, we assessed the practical significance of our results by associating the detection and classification of differential combinatorial patterns from called peaks with gene expression (Section \ref{sec:predpostprob}; Web Appendix D1).

We compared the genome-wide performance of epigraHMM with ChIPComp, csaw, DiffBind, diffReps, RSEG, and THOR. For two-step DPCs, ChIPComp and DiffBind, we used peak calls from MACS2 as candidate peaks \citep{zhang2008model}. Methods were compared in terms of the coverage of differentially enriched genes, the number and average size of differential peak calls, the $\log_{2}$ fold change (LFC) and Spearman correlation of $\log_{2}$ reads counts mapped onto differential peak calls, the computation time, and maximum memory utilized. Metrics for sensitivity and specificity were defined on the window level and based on the coverage of differentially enriched genes by called peaks (Web Appendix D1). Read counts were computed using non-overlapping windows of 500bp and 250bp for broad and short marks, respectively (see Web Appendix D2 for a discussion on window size). Results presented here pertain to the analysis of cell lines Helas3 and Hepg2 (see Web Appendix D3 for additional results). Analyses were adjusted by input control samples in methods that are designed to do so (ChIPComp, DiffBind, diffReps, and THOR). We did not observe a significant improvement in performance by accounting for input control effect or GC-content bias with epigraHMM and did not adjust our analyses for these effects (Web Appendix B1). 

\subsection{Analysis of ChIP-seq Data From Broad Marks}
\label{sec:broad}
We compared methods regarding the coverage of differentially enriched genes by differential peak calls for H3K36me3, an epigenomic mark associated with transcribed genes. Following \cite{ji2013differential}, in which the authors define true differential binding sites from LFCs of ChIP-seq counts to assess model sensitivity (Figure 2E in \cite{ji2013differential}), we defined a set of protein coding genes exhibiting $|\text{LFC}|>2$ of ChIP-seq counts between Helas3 and Hepg2 cell lines as true differential binding sites (see Web Appendix D1 for results using different thresholding values). Median ratio normalization of counts was performed on protein-coding genes after excluding those with total counts under the $25^{th}$ percentile of the distribution. Figure~\ref{fig:fig_broad}A shows receiver operating characteristic (ROC) curves for various methods and different nominal FDR levels for H3K36me3 differential peak calls. Observed true (false) positive rates were computed on the window-level as the proportion of windows called as differential out of all windows associated (not associated) with differentially enriched genes (Web Appendix D1). Overall, epigraHMM outperformed other DPCs by covering most of the differentially enriched genes while calling a limited number of false positive peaks. 

DiffBind, which has been shown to be dependent on the set of candidate peaks \citep{lun2015csaw}, called the shortest peaks overall (Figures \ref{fig:fig_broad}B and \ref{fig:fig_broad}C) and ChIPComp, which has been shown to perform best in scenarios with short marks \citep{steinhauser2016comprehensive}, appeared to have limited performance in broad marks such as H3K27me3 and EZH2 (Figures \ref{fig:fig_broad}D and \ref{fig:fig_broad}F). epigraHMM and RSEG, two HMM-based methods, called broader differential peaks and exhibited better sensitivity than other methods. Yet, differential peaks from RSEG often did not correspond to observed differential enrichment, a fact that explains its high observed FDR and agrees with our simulations (Figures \ref{fig:fig_broad}C and \ref{fig:fig_simulation_csaw}) and others \citep{starmer2016detecting}. Our HMM-based approach with a non-linear normalization via model offsets allowed us to maintain a low observed FDR and a higher sensitivity than other DPCs. Overall, epigraHMM was among the most efficient algorithms due to our computational scheme, taking approximately 30 minutes to analyze genome-wide data (Figure~\ref{fig:fig_broad}E).

\begin{figure}
\begin{center}
\includegraphics[scale = 0.65]{../../Main/Figure_Broad/Figure_Broad.pdf}
\end{center}
\caption{Analysis of broad ENCODE data. (A): ROC curves of DPCs covering differentially enriched genes ($|\text{LFC}|>2$ of ChIP-seq counts) for H3K36me3. (C): sensitivity (y-axis) and average H3K36me3 differential peak size (kbp; x-axis) of various methods under different nominal FDR thresholds (observed FDR annotated next to data points). (B), (D), and (F): example of peak calls from H3K36me3, H3K27me3, and EZH2, respectively, under a nominal FDR control of 0.05. Posterior probabilities of the HMM differential state are shown at the bottom of each panel. (E): computing time and peak memory usage of genome-wide analysis from various methods. This figure appears in color in the electronic version of this article.
\label{fig:fig_broad}}
\end{figure}

\subsection{Analysis of ChIP-seq Data From Short Marks}
\label{sec:sharp}

We further evaluated the performance epigraHMM on ChIP-seq samples characterized by short peaks (CTCF, H3K4me3, and H3K27ac). The goal of our analysis was to assess whether epigraHMM was flexible to different types of data and still able to call short differential regions of enrichment. Here, differential peaks are usually observed in isolated genomic regions and exhibit a high SNR. It has been shown that certain HMM-based algorithms, including RSEG, have low accuracy with such short marks \citep{hocking2016optimizing}. 

We calculated the LFC and the Spearman correlation between cell lines Helas3 and Hepg2 of ChIP-seq counts mapped onto differential peaks called by each method. Median ratio normalization of counts was performed prior to computing such metrics. Here, ideal methods would show high absolute LFC and negative correlation between read counts mapped onto their differential peaks. We observed that other HMM-based algorithms, RSEG and THOR, were among those with the lowest absolute LFC among all methods, which confirms their sub optimal performance in the scenario of short peaks (Figures \ref{fig:fig_sharp}A and \ref{fig:fig_sharp}C). ChIPComp, DiffBind, and csaw, were among those with the best performance overall, as their differential peak calls had the highest absolute LFCs and the lowest Spearman correlation of counts between cell lines. Overall, epigraHMM performed comparably to these methods that are known to perform best in short epigenomic marks while calling differential peaks in less than 1.5 hour (Figures \ref{fig:fig_sharp}B and \ref{fig:fig_sharp}D-\ref{fig:fig_sharp}F; \cite{steinhauser2016comprehensive}).


\begin{figure}
\begin{center}
\includegraphics[scale = 0.65]{../../Main/Figure_Narrow/Figure_Narrow.pdf}
\end{center}
\caption{Analysis of short ENCODE data. (A) and (C): median LFC and correlation between cell lines of ChIP-seq counts from differential peaks for CTCF and H3K4me3, respectively. (B), (D), and (F): example of peak calls from CTCF, H3K4me3, and H3K27ac, respectively. Posterior probabilities of the HMM differential state are shown at the bottom of each panel. (E): computing time and peak memory usage of genome-wide analysis from various methods. Results are shown under a nominal FDR control of 0.05. This figure appears in color in the electronic version of this article.
\label{fig:fig_sharp}}
\end{figure}

\subsection{Genomic Segmentation and Classification of Chromatin States}
\label{sec:predpostprob}

We analyzed data from the cell line Helas3 to segment its genome regarding the joint activity of H3K36me3, H3K27me3, and EZH2. We considered each mark as a separate experimental condition ($G=3$) and jointly classified local chromatin states based upon the presence or absence of enrichment from each mark. Assuming that EZH2 catalyzes the methylation of H3K27me3, a repressive mark, and H3K36me3 associates with transcribed genes (Section \ref{sec:background}), we expected consensus enrichment regions to be rare and differential regions to be mostly represented by either transcribed chromatin states (enrichment of H3K36me3 alone) or repressed chromatin states (co-enrichment for H3K27me3 and EZH2). The analyses presented here highlight the applicability of our method in the context of genomic segmentation \citep{ernst2012chromhmm}, a distinct problem not tackled by current DPCs.

First, we observed that while the majority of genomic regions exhibited no enrichment for any of the analyzed marks, regions of consensus enrichment were rare and covered only 2\% of the genome (Figure~\ref{fig:fig_multimarks}A), as expected. Consensus background and differential regions mostly covered intergenic and protein-coding genic regions, respectively. In fact, differential genomic windows were mostly representing either transcribed chromatin states or repressed chromatin states (Figure~\ref{fig:fig_multimarks}B). All differential combinatorial patterns expected to be rare had estimated mixture model proportions less than 0.05, which were then pruned using a model selection scheme via BIC (Figure~\ref{fig:fig_multimarks}C; Web Appendix B2).

\begin{figure}
\begin{center}
\includegraphics[scale = 0.7]{../../Main/Figure_MultiMarks/Figure_MultiMarks.pdf}
\end{center}
\caption{Genomic chromatin state segmentation and classification. (A): distribution of base pairs (y-axis) and the Viterbi sequence of states (x-axis). (B): estimated mixture probabilities and the associated differential combinatorial patterns. (C): choice of best model with two mixture components via BIC. (D): density estimate from expression of genes intersecting differential peaks associated with the enrichment of H3K36me3 alone or the enrichment of H3K27me3 and EZH2 in consensus. (E): cycle-regulated genes and their associated average mixture posterior probabilities associated with differential H3K36me3 enrichment pattern of overlapping differential peaks. (F): example of a genomic region with differential peaks and genes, colored according to their classification and expression levels, respectively. This figure appears in color in the electronic version of this article.
\label{fig:fig_multimarks}}
\end{figure}

To assess the biological significance of our results, we associated the distribution of transcripts per million (TPM, from matching RNA-seq experiments) of protein-coding genes with combinatorial patterns of overlapping differential peaks. To assign combinatorial patterns to differential peaks, we chose the combination pertaining to the most prevalent mixture component across windows by using the maximum estimated mixture model posterior probability, $Pr(W_{jl}=1|Z_{j}=2,\mb{y}_{..j},\mb{x};\bs{\Psi}^{(t)})$, $j=1,\ldots,M$. Genes associated with transcribed chromatin states had a significantly higher distribution of TPM counts than genes associated with repressed chromatin states (Figure~\ref{fig:fig_multimarks}D). In fact, nearly all known Helas3 cycle-regulated genes (therefore genes expressed at some stage of the cell life cycle; \cite{whitfield2002identification}) were associated with high average mixture posterior probabilities of differential enrichment for H3K36me3 alone (Figure~\ref{fig:fig_multimarks}E).

The expression of such genes are known to correlate with proliferative states of tumors and their study may help characterize their role in cancer \citep{whitfield2002identification}. Differential peaks called by epigraHMM, and their assigned differential combinatorial pattern, often agreed with the biological roles as well as the expression levels of associated genes (Figure~\ref{fig:fig_multimarks}F; results from ChromHMM with 3 states for comparison). epigraHMM offers the benefit of simultaneously detecting differential peaks and classifying differential combinatorial patterns of enrichment even in the context of genomic segmentation of highly diverse epigenomic marks (Web Appendix D4). By using the BIC for model selection, one can choose the number of biologically relevant mixture components to be included in the model, a task that may not be as straightforward in methods such as ChromHMM (Web Appendix B2).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Discussion}
\label{sec:discussion}

We presented a flexible and efficient statistical model designed to call differential regions of enrichment from ChIP-seq experiments with multiple replicates and multiple conditions. Our model has three main advantages over current methods tailored for differential peak detection. First, it uses an HMM-based approach that accounts for the local dependency of ChIP-seq counts and is able to precisely detect broad and short differential regions of enrichment. Second, it utilizes a GLM-based framework with model offsets that accounts for non-linear biases and other potential factors such as GC-content bias or input control samples. Our efficient implementation of the RCEM algorithm led to genome-wide analyses of ChIP-seq data under a computational time comparable to some of the fastest current algorithms. Our results highlight the importance of the inclusion of technical/biological replicates in the analysis of epigenomic data. Lastly, our method allows the simultaneous detection and classification of differential combinatorial patterns of enrichment from its embedded mixture model and the associated posterior probabilities under any number of conditions. When certain differential combinatorial patterns are rare, our model selection scheme via BIC provides means of pruning specific patterns of interest (Web Appendix B2). % Data resulting from these assays are similar in nature but aim to identify regions that alternatively are not bound to proteins. % Such a tool could trigger, for instance, new insights on the field of epigenomics as investigators are often interested in making sense of the roles of epigenomic marks on the human genome. % Analysis of datasets from ther epigenomic assays, such as ATAC-seq, could also benefit from the presented method.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\backmatter
% \section*{Acknowledgements}
% The authors thank Professor A. Sen for some helpful suggestions,
% Dr C. R. Rangarajan for a critical reading of the original version of the
% paper, and an anonymous referee for very useful comments that improved
% the presentation of the paper.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\bibliographystyle{biom}\bibliography{bibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Supporting Information}
Web Appendices A-D, referenced in Sections 1-6, and a link to the implemented software are available with this paper at the Biometrics website on Wiley Online Library. MACS2 peak caller is available at https://github.com/macs3-project/MACS. epigraHMM has been implemented as an R package and it is available for download at https://github.com/plbaldoni/epigraHMM.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\label{lastpage}
\end{document}

